{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e6ed8d18-5cb0-4785-999d-7ca81c82a03f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# **PURE CRN model of TX-TL models for MGapt**\n",
    "\n",
    "In this notebook, we built on the proposed chemical reaction network transcription model for PURE protein synthesis by including the translation reaction without protein production. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8888a48c-32f1-41cc-a543-9fd6ed77957d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Importing required packages and definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0b4a8a3e-9b13-40cc-ba27-54d34521c6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biocrnpyler\n",
    "from biocrnpyler import *\n",
    "from biocrnpyler.component import Component\n",
    "from biocrnpyler.chemical_reaction_network import Species, Reaction, ChemicalReactionNetwork\n",
    "from biocrnpyler.mechanism import Mechanism\n",
    "from biocrnpyler.reaction import Reaction\n",
    "from biocrnpyler.species import Complex, Species, WeightedSpecies\n",
    "\n",
    "import bioscrape\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pylab as plt\n",
    "from scipy import stats\n",
    "\n",
    "import libsbml\n",
    "import time\n",
    "from typing import List, Union\n",
    "import csv\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import random \n",
    "import datetime\n",
    "\n",
    "#Get directory\n",
    "import os\n",
    "directory = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bf265690-9480-424e-9f3a-ded3bfff3b18",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_custom_plot(title_text, x_max=8,y_max=2, yname='MGapt (μM)'):\n",
    "    custom_plot = figure(\n",
    "        toolbar_location='right',\n",
    "        outline_line_color=None,\n",
    "        min_border_right=10,\n",
    "        height=400,\n",
    "        width=400,\n",
    "    )\n",
    "\n",
    "    custom_plot.title.text = title_text\n",
    "    custom_plot.xaxis.axis_label = 'Time (hours)'\n",
    "    custom_plot.yaxis.axis_label = yname\n",
    "    custom_plot.y_range = Range1d(0, y_max)\n",
    "    custom_plot.x_range = Range1d(0, x_max)\n",
    "    custom_plot.outline_line_color = None\n",
    "\n",
    "    # custom_plot.yaxis\n",
    "    custom_plot.ygrid.visible = False\n",
    "    custom_plot.yaxis.axis_label_text_font_size = '15pt'\n",
    "    custom_plot.yaxis.major_label_text_font_size = '15pt'\n",
    "    custom_plot.yaxis.major_label_text_font = 'Work Sans'\n",
    "    custom_plot.yaxis.axis_label_standoff = 15\n",
    "    custom_plot.yaxis.axis_label_text_font_style = 'normal'\n",
    "\n",
    "    # custom_plot.xaxis\n",
    "    custom_plot.xgrid.visible = False\n",
    "    custom_plot.xaxis.axis_label_text_font_size = '15pt'\n",
    "    custom_plot.xaxis.major_label_text_font_size = '15pt'\n",
    "    custom_plot.xaxis.major_label_text_font = 'Work Sans'\n",
    "    custom_plot.xaxis.axis_label_standoff = 15\n",
    "    custom_plot.xaxis.axis_label_text_font_style = 'normal'\n",
    "\n",
    "    # custom_plot.title\n",
    "    custom_plot.title.text_font_size = '18pt'\n",
    "    custom_plot.title.align = 'left'\n",
    "    custom_plot.title.offset = -70.0\n",
    "\n",
    "    return custom_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0c54cd-1f16-4f13-97c2-44903bc054cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Bokeh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09c4207c-5d5f-4830-900c-a1961714c5df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.0.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.0.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import bokeh.io\n",
    "import bokeh.plotting\n",
    "bokeh.io.output_notebook()\n",
    "from bokeh.themes import Theme\n",
    "\n",
    "# Modules needed from Bokeh.\n",
    "from bokeh.io import output_file, show\n",
    "from bokeh.plotting import gridplot,figure\n",
    "from bokeh.models import LinearAxis, Range1d\n",
    "\n",
    "colors2 = bokeh.palettes.Colorblind[8]\n",
    "\n",
    "try:\n",
    "    import dnaplotlib as dpl\n",
    "    dpl_enabled = True\n",
    "except (ModuleNotFoundError,ImportError) as e:\n",
    "    dpl_enabled = False\n",
    "    \n",
    "from bokeh.io import export_png"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e33caf68-5df2-488e-b9fb-2e7eef92ced9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e717c65a-5815-41a3-bb0b-0e77f2c14c21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Definition to un-nest lists that will be created by models\n",
    "def flatten(t):\n",
    "    return [item for sublist in t for item in sublist]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ce4a5d3-06a8-41ff-844c-4137424efb4a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# User defined transcribed unit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4eba3b3c-844f-427f-bc2f-9153cd502a0e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#User will copy the sequence bwteen end of promoter and start of terminator of DNA strain\n",
    "#mGapt sequence\n",
    "dna_seq='GGATCCCGACTGGCGAGAGCCAGGTAACGAATGGATCTCGAGCCTTAGGAGATCCGGCTGCTAACAAAGCCCGAAAGGAAGCTGAGTTGGCTGCTGCCACCGCTGAGCAATAA'\n",
    "#mGapt_UTR1_deGFP \n",
    "# dna_seq= 'GGGATCCCGACTGGCGAGAGCCAGGTAACGAATGGATCCAATAATTTTGTTTAACTTTAAGAAGGAGATATACCATGGAGCTTTTCACTGGCGTTGTTCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCTAACTCGAGCCTTAGGAGATCCGGCTGCTAACAAAGCCCGAAAGGAAGCTGAGTTG'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63aa5c2d-1c39-42c3-aad6-0aac2bde0292",
   "metadata": {
    "tags": []
   },
   "source": [
    "# PURE TX Model in BioCRNpyler\n",
    "This model is based on the TX-TL from Tuza *et al.* [1] originally built for cell lysate protein expression. This section of code requires a DNA sequence input from the user. If transcription is not desired, an amino acid list can also serve as the input and this section can be omitted. The steps of the TX are split into 3 subsections: species and reactions based on nucleotide needed (1), generates complement sequence (2), and species and reactions of growing mRNA strand (3). \n",
    "\n",
    "<ol>\n",
    "  <li>Non-nucleic acid Species and Reactions</li>\n",
    "  <li>RNA sequence</li>\n",
    "  <li> Addition of nucleotide and individual reactions for the desired DNA </li>\n",
    "</ol>\n",
    " \n",
    "### Citation:\n",
    "[1] Zoltán A. Tuza, Dan Siegal-Gaskins, Jongmin Kim, and Gábor Szederkényi. Analysis-based parameter estimation of an *in vitro* transcription-translation system. In Eur. Control Conf. 2015, page 1554–1560, 2015. ([PNAS, 2015](https://ieeexplore.ieee.org/document/7330760))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b7f35472-e4ef-497d-9601-f93a00d29fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = bokeh.plotting.figure(toolbar_location='right',\n",
    "    outline_line_color= None,\n",
    "    min_border_right=10,\n",
    "    height=400,\n",
    "    width=600,)\n",
    "\n",
    "p1.title.text = \"A) Simulation Results of mRNA\"\n",
    "p1.xaxis.axis_label = 'Time (hours)'\n",
    "p1.yaxis.axis_label = 'mRNA (μM)'\n",
    "p1.outline_line_color=None\n",
    "\n",
    "# p1.yaxis\n",
    "p1.ygrid.visible = False\n",
    "p1.yaxis.axis_label_text_font_size='15pt' \n",
    "p1.yaxis.major_label_text_font_size = '15pt'\n",
    "p1.yaxis.major_label_text_font='Work Sans'\n",
    "p1.yaxis.axis_label_standoff=15\n",
    "p1.yaxis.axis_label_text_font_style='normal' \n",
    "\n",
    "# p1.xaxis\n",
    "p1.xgrid.visible = False\n",
    "p1.xaxis.axis_label_text_font_size='15pt' \n",
    "p1.xaxis.major_label_text_font_size = '15pt'\n",
    "p1.xaxis.major_label_text_font='Work Sans'\n",
    "p1.xaxis.axis_label_standoff=15\n",
    "p1.xaxis.axis_label_text_font_style='normal' \n",
    "\n",
    "# p1.title\n",
    "p1.title.text_font_size= '18pt' \n",
    "p1.title.align= 'left'\n",
    "p1.title.offset=-65.0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18f2555e-15d0-4737-bec0-199566b8d295",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Non-nucleic acid Species and Reactions\n",
    "<!--  -->\n",
    "Creates all the species and reaction around the nucleotide sequences in the given transcribed regions, common protein and small molecules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a69b5dc7-635c-41f4-a874-5d37a25157e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Final parameters based on MGapt-TXonly fine tuning with B4 data\n",
    "k_rnapbF1 = 9.406062\n",
    "k_rnapbF2 = 5.080765\n",
    "k_rnapbF3 = 10.656306\n",
    "k_start= 7.463710\n",
    "k_ntpbound=  2.676172\n",
    "k_ntpadd  = 100.066259\n",
    "k_ntpdis =  1306.617911\n",
    "k_term  =  45.998466"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59de9252-1334-4c91-8bfa-cbd03eb6f52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Species\n",
    "#Define species needed for TX \n",
    "T7RNAP = Species('T7RNAP') #ActivePolymerase\n",
    "DNA = Species(\"DNA\") \n",
    "mRNA_i=Species('mRNA_i') # mRNA intermediate species\n",
    "mRNA_t=Species('mRNA_t') # mRNA intermediate species\n",
    "MGapt=Species('MGapt') # mRNA species used to track total mRNA production\n",
    "T7RNAP_bound = Species(\"T7RNAP_bound\") \n",
    "T7RNAP_bound_GTP=Species('T7RNAP_bound_GTP')\n",
    "T7RNAP_bound_GDP_PO4= Species('T7RNAP_bound_GDP_PO4')\n",
    "\n",
    "#Nucleotide bases that are \"active\"\n",
    "ATP=Species('ATP')\n",
    "GTP=Species('GTP')\n",
    "CTP=Species('CTP')\n",
    "UTP=Species('UTP')\n",
    "\n",
    "#Other small molecules\n",
    "GDP=Species('GDP')\n",
    "PPi=Species('PPi')\n",
    "PO4=Species('PO4')\n",
    "\n",
    "#Define placing all species in one array \n",
    "Species_general= [T7RNAP, DNA, mRNA_i, MGapt, mRNA_t,\n",
    "                  ATP, GTP, CTP, UTP,PPi, PO4,                   \n",
    "                  T7RNAP_bound_GTP, T7RNAP_bound, \n",
    "                  T7RNAP_bound_GDP_PO4, T7RNAP_bound_GTP,]\n",
    "\n",
    "#Define Reactions needed for TX regardless of the sequence defined\n",
    "RXN_general= [\n",
    "    #Binding of RNAP and beginning of transcription\n",
    "    Reaction.from_massaction([T7RNAP, DNA, GTP], [T7RNAP_bound_GTP], k_forward=k_rnapbF1),\n",
    "    Reaction.from_massaction([T7RNAP_bound_GTP], [T7RNAP_bound_GDP_PO4], k_forward=k_rnapbF2),\n",
    "    Reaction.from_massaction([T7RNAP_bound_GDP_PO4], [T7RNAP_bound,GDP, PO4], k_forward=k_rnapbF3),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da4e9e7-daed-48ff-8c67-45533284cb2f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Transcribing DNA sequence to mRNA sequence\n",
    "\n",
    "Transcribed the DNA input into RNA nucleotides. Output is a list of the nucleotides used to make the mRNA. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5505c6a1-3597-4e4b-aee1-0c42c8491921",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Transcribed given DNA sequence into mRNA sequence\n",
    "def getTranscript(dna):\n",
    "    transcript = []\n",
    "    for bp in dna:\n",
    "        if bp == 'A':\n",
    "            tx = ATP\n",
    "        elif bp == 'T':\n",
    "            tx =UTP\n",
    "        elif bp == 'G':\n",
    "            tx =GTP\n",
    "        elif bp == 'C':\n",
    "            tx =CTP\n",
    "        else:\n",
    "            print('Non-standard nucleotides')\n",
    "            \n",
    "        transcript.append(tx)\n",
    "\n",
    "    return transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "680a7c3c-c939-4c46-a3c5-5591a6da573f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Translating RNA sequence from DNA sequence\n",
    "rna_seq = getTranscript(dna_seq)\n",
    "# print(rna_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3758ce3-3421-4b43-9319-595305886a6b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Addition of nucleotide and individual reactions for the desired DNA\n",
    "\n",
    "Individually adds each nucleotide of the growing mRNA strand. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fc7fcd67-0331-4b1a-986d-a06d33ef2f45",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Start iternations to loop over mRNA sequence\n",
    "nt_len = len(rna_seq)\n",
    "mrna_length='0000'\n",
    "rxn_list=[]\n",
    "species_list=[]\n",
    "\n",
    "for L in [l for l in range(len(rna_seq))]:\n",
    "    ntp = rna_seq[L]\n",
    "    mrna0 = mrna_length[:-len(str(L))]+str(L) #starting mRNA\n",
    "    mrnaG = mrna_length[:-len(str(L+1))]+str(L+1) #Growing mRNA\n",
    "    \n",
    "    #Initiation of mRNA strand \n",
    "    if L == 0:       \n",
    "        T7RNAP_bound_mRNAmrna0 = Species('T7RNAP_bound_mRNA'+mrna0)\n",
    "        T7RNAP_bound_mRNAmrna0_ntp = Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))\n",
    "        T7RNAP_bound_mRNAmrnaG_nmp_PPi = Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')\n",
    "        T7RNAP_bound_mRNAmrnaG = Species('T7RNAP_bound_mRNA'+mrnaG)\n",
    "        \n",
    "        species= [T7RNAP_bound_mRNAmrna0, T7RNAP_bound_mRNAmrna0_ntp, T7RNAP_bound_mRNAmrnaG_nmp_PPi, T7RNAP_bound_mRNAmrnaG]\n",
    "        \n",
    "        rxns=[Reaction.from_massaction([T7RNAP_bound],[Species('T7RNAP_bound_mRNA'+mrna0)],k_forward=k_start), #can add DNA, DNA\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0), ntp], [Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))], k_forward=k_ntpbound),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))],[Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],k_forward=k_ntpadd),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],[Species('T7RNAP_bound_mRNA'+mrnaG), PPi],k_forward=k_ntpdis),] \n",
    "     \n",
    "    #Elongation of mRNA strand  \n",
    "    elif L<nt_len-1:\n",
    "        T7RNAP_bound_mRNAmrna0_ntp = Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))\n",
    "        T7RNAP_bound_mRNAmrnaG_nmp_PPi = Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')\n",
    "        T7RNAP_bound_mRNAmrnaG = Species('T7RNAP_bound_mRNA'+mrnaG)\n",
    "\n",
    "        species= [T7RNAP_bound_mRNAmrna0_ntp, T7RNAP_bound_mRNAmrnaG_nmp_PPi, T7RNAP_bound_mRNAmrnaG]\n",
    "            \n",
    "        if L==38:\n",
    "            rxns=[Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0), ntp], [Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))], k_forward=k_ntpbound),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))],[Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],k_forward=k_ntpadd),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],[Species('T7RNAP_bound_mRNA'+mrnaG), PPi,MGapt],k_forward=k_ntpdis),] \n",
    "        else:\n",
    "            rxns=[Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0), ntp], [Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))], k_forward=k_ntpbound),\n",
    "                  Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))],[Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],k_forward=k_ntpadd),\n",
    "                  Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],[Species('T7RNAP_bound_mRNA'+mrnaG), PPi,],k_forward=k_ntpdis),] \n",
    "        \n",
    "    #Termination of mRNA strand      \n",
    "    else:\n",
    "        T7RNAP_bound_mRNAmrna0_ntp = Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))\n",
    "        T7RNAP_bound_mRNAmrnaG_nmp_PPi = Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')\n",
    "        T7RNAP_bound_mRNAmrnaG = Species('T7RNAP_bound_mRNA'+mrnaG)\n",
    "        \n",
    "        species= [T7RNAP_bound_mRNAmrna0_ntp, T7RNAP_bound_mRNAmrnaG_nmp_PPi, T7RNAP_bound_mRNAmrnaG,]\n",
    "                   \n",
    "        rxns=[Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0), ntp], [Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))], k_forward=k_ntpbound),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrna0+'_'+ str(ntp))],[Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],k_forward=k_ntpadd),  \n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrnaG+'_PPi')],[Species('T7RNAP_bound_mRNA'+mrnaG), PPi],k_forward=k_ntpdis),\n",
    "              Reaction.from_massaction([Species('T7RNAP_bound_mRNA'+mrnaG),],[T7RNAP, DNA, mRNA_i,mRNA_t],k_forward=k_term)]#\n",
    "        \n",
    "    #Appending each reaction and species to the one before       \n",
    "    rxn_list.append(rxns)\n",
    "    species_list.append(species)\n",
    "\n",
    "#Flattening to th nested lists \n",
    "species_list = flatten(species_list)\n",
    "rxn_list= flatten(rxn_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "836cdc1b-339a-447a-b8d3-2669491ea87d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Compiling TX CRN (CRN_TX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "81c939c0-eece-4441-b13a-cfceef5c42c4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(344, 355)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#List of all species and reactions\n",
    "All_species_TX = flatten([Species_general, species_list])\n",
    "All_rxn_TX = flatten([RXN_general, rxn_list])\n",
    "\n",
    "#Buliding the CRN_TX and saving as a SBML file\n",
    "CRN_TX = ChemicalReactionNetwork(species = All_species_TX, reactions = All_rxn_TX)\n",
    "\n",
    "len(CRN_TX.reactions),len(CRN_TX.species)\n",
    "#MGapt_only:  (345, 356)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e37b26a6-fdb9-45b1-a21f-e4f5629029b1",
   "metadata": {
    "tags": []
   },
   "source": [
    "# PURE TL Model in BioCRNpyler\n",
    "\n",
    "This model is based on the PURE TL models from Matsuura *et al.* [2,3] originally modeled in MATLAB. This section of code uses the amino acids list derived from the DNA sequence input to generate all species and reactions associated with desired protein. If transcription is not wanted, an amino acid list can also serve as the input. The PURE TL does not account for multiple tRNAs coding for the same amino acid.\n",
    "\n",
    "For order and simplicity, the reactions were broken up into 6 subsections: species and reactions based on amino acids needed (1-2) and species and reactions of growing peptide (3-6).\n",
    "\n",
    "<ol>\n",
    "  <li>Non-nucleic acid Species and Reactions</li>\n",
    "  <li>Addition of AA individual reactions</li>\n",
    "  <li>Met Species and Reaction (Start codon)</li>\n",
    "  <li>Elongation</li>\n",
    "  <li>Termination (End codon)</li>\n",
    "  <li>Removal of Ribosome</li>\n",
    "</ol>\n",
    "\n",
    "### Citation:\n",
    "[2] Matsuura, T., Tanimura, N., Hosoda, K., and Shimizu, Y. Reaction dynamics analysis of a reconstituted *Escherichia coli* protein translation system by computational modeling. PNAS. 2017, 114, 8. ([PNAS, 2017](https://www.pnas.org/doi/10.1073/pnas.1615351114)) \\\n",
    "[3] Matsuura, T., Hosoda, K., and Shimizu, Y. Robustness of a reconstituted *Escherichia coli* protein translation system analyzed by computational modeling. ACS Synth. Biol. 2018, 7, 8, 1964–1972. ([ACS Synth. Biol., 2018](https://pubs.acs.org/doi/10.1021/acssynbio.8b00228))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4d80c97-b070-4df3-b561-dfeafd46361d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define the directory where parameters and initial conditions are saved as a CSV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "502dfe21-1763-4faf-999c-0becb549297f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Directory and file for the reaction rates\n",
    "filename_parameters = '/Data_files/fMGG_synthesis_parameters_CRN.csv'\n",
    "\n",
    "with open(directory + filename_parameters, mode='r') as infile:\n",
    "    reader = csv.reader(infile)\n",
    "    rxn_k= {rows[0]:float(rows[1]) for rows in reader}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4af53b07-b528-4085-a761-af7eb4918f1e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Translating DNA sequence to peptide amino acid list\n",
    "Converting DNA sequence into amino acid list for translation model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed914bd7-251b-4ddf-b602-a621f0ee62ad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#DNA code chart for coding domain\n",
    "def translate(seq): \n",
    "    table = {\n",
    "        'ATA':'Ile', 'ATC':'Ile', 'ATT':'Ile', 'ATG':'Met',\n",
    "        'ACA':'Thr', 'ACC':'Thr', 'ACG':'Thr', 'ACT':'Thr',\n",
    "        'AAC':'Asn', 'AAT':'Asn', 'AAA':'Lys', 'AAG':'Lys',\n",
    "        'AGC':'Ser', 'AGT':'Ser', 'AGA':'Arg', 'AGG':'Arg',                 \n",
    "        'CTA':'Leu', 'CTC':'Leu', 'CTG':'Leu', 'CTT':'Leu',\n",
    "        'CCA':'Pro', 'CCC':'Pro', 'CCG':'Pro', 'CCT':'Pro',\n",
    "        'CAC':'His', 'CAT':'His', 'CAA':'Gln', 'CAG':'Gln',\n",
    "        'CGA':'Arg', 'CGC':'Arg', 'CGG':'Arg', 'CGT':'Arg',\n",
    "        'GTA':'Val', 'GTC':'Val', 'GTG':'Val', 'GTT':'Val',\n",
    "        'GCA':'Ala', 'GCC':'Ala', 'GCG':'Ala', 'GCT':'Ala',\n",
    "        'GAC':'Asp', 'GAT':'Asp', 'GAA':'Glu', 'GAG':'Glu',\n",
    "        'GGA':'Gly', 'GGC':'Gly', 'GGG':'Gly', 'GGT':'Gly',\n",
    "        'TCA':'Ser', 'TCC':'Ser', 'TCG':'Ser', 'TCT':'Ser',\n",
    "        'TTC':'Phe', 'TTT':'Phe', 'TTA':'Leu', 'TTG':'Leu',\n",
    "        'TAC':'Tyr', 'TAT':'Tyr', 'TAA':'_', 'TAG':'_',\n",
    "        'TGC':'Cys', 'TGT':'Cys', 'TGA':'_', 'TGG':'Trp', }\n",
    "    \n",
    "    protein =[]\n",
    "    if len(seq)%3 == 0:\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codon = seq[i:i + 3]\n",
    "            if table[codon] != '_':\n",
    "                protein+= [table[codon]]\n",
    "            else:\n",
    "                break\n",
    "    return protein\n",
    "\n",
    "#Translation of CDS\n",
    "def coding_protein(dna, start=0):\n",
    "    for bp in range(start,len(dna)-2):\n",
    "        bp_aa = dna_seq[bp:bp+3]\n",
    "        \n",
    "        if bp_aa=='ATG':\n",
    "            start = bp\n",
    "            remainder= len(dna[bp:]) % 3\n",
    "            \n",
    "            if remainder ==0:\n",
    "                CDS= dna\n",
    "            else:\n",
    "                CDS= dna[bp:-remainder]\n",
    "            \n",
    "            translation= translate(CDS)\n",
    "            \n",
    "            if len(translation)>0:\n",
    "                return translation\n",
    "                break\n",
    "            \n",
    "        if bp == len(dna)-2:\n",
    "            print('No start codon found')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "48515ad2-6c9a-4d95-b4b1-1d648ea8088e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Translating protein sequence from DNA sequence\n",
    "# protein=coding_protein(dna_seq)\n",
    "protein=['Met','Ala', 'Arg', 'Asn', 'Asp', \n",
    "         'Cys', 'Gln', 'Glu', 'Gly', 'His', \n",
    "         'Ile','Leu', 'Lys', 'Phe', 'Pro',\n",
    "          'Ser', 'Thr', 'Trp', 'Tyr', 'Val']\n",
    "# print(protein,len(protein))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fde45f1-19ab-478f-b981-837b8e844109",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Non-nucleic acid Species and Reactions\n",
    "\n",
    "Creates all the species and reaction around the amino acids in the given chain, common protein and small molecules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1264a331-b3b9-4839-b25d-a5b08b384fc2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Makes a list of the aa needed in the given amino acid chain, removing any duplicates.\n",
    "AA= list(set(protein[1:],)) #removes any duplicated amino acids\n",
    "\n",
    "#Makes empty array for species and reactions\n",
    "list_of_reactions = []\n",
    "list_species_aa=[]\n",
    "\n",
    "#Reference lists for aa and respective codon.\n",
    "list_AA =['Ala', 'Arg', 'Asn', 'Asp', 'Cys', \n",
    "          'Gln', 'Glu', 'Gly', 'His', 'Ile', \n",
    "          'Leu', 'Lys', 'Met', 'Phe', 'Pro',\n",
    "          'Ser', 'Thr', 'Trp', 'Tyr', 'Val']\n",
    "list_condon =['AGU', 'CCG', 'AUU','AUC','ACA',\n",
    "             'UUG','UUC','GCC','AUG','AAU',\n",
    "             'CAG','CUU','CAU','GAA','AGG',\n",
    "             'GCU','AGU','CCA','AUA','CAC',]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d9333a-73c6-40c3-bb01-f6f8fc83b93c",
   "metadata": {
    "tags": []
   },
   "source": [
    "### General proteins and small molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ea8bdc06-7c0d-40cc-9f9e-da096156c9a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Make species for general proteins and small molecules\n",
    "mRNA=Species('mRNA')\n",
    "ADP=Species('ADP')\n",
    "# GDP=Species('GDP')\n",
    "AMP=Species('AMP')\n",
    "GMP=Species('GMP')\n",
    "CMP=Species('CMP')\n",
    "UMP=Species('UMP')\n",
    "\n",
    "CK = Species('CK')\n",
    "CK_ADP = Species('CK_ADP')\n",
    "CK_ATP = Species('CK_ATP')\n",
    "CK_CP = Species('CK_CP')\n",
    "CK_CP_ADP = Species('CK_CP_ADP')\n",
    "CK_Cr = Species('CK_Cr')\n",
    "CK_Cr_ATP = Species('CK_Cr_ATP')\n",
    "\n",
    "CK_degraded = Species('CK_degraded')\n",
    "CP = Species('CP')\n",
    "Cr = Species('Cr')\n",
    "EFG = Species('EFG')\n",
    "EFG_degraded = Species('EFG_degraded')\n",
    "EFG_GDP = Species('EFG_GDP')\n",
    "EFG_GTP = Species('EFG_GTP')\n",
    "EFTs = Species('EFTs')\n",
    "EFTs_degraded = Species('EFTs_degraded')\n",
    "EFTu = Species('EFTu')\n",
    "\n",
    "EFTu_degraded = Species('EFTu_degraded')\n",
    "EFTu_EFTs = Species('EFTu_EFTs')\n",
    "EFTu_GDP = Species('EFTu_GDP')\n",
    "EFTu_GDP_EFTs = Species('EFTu_GDP_EFTs')\n",
    "EFTu_GTP = Species('EFTu_GTP')\n",
    "EFTu_GTP_EFTs = Species('EFTu_GTP_EFTs')\n",
    "FD = Species('FD')\n",
    "\n",
    "IF1 = Species('IF1')\n",
    "IF1_degraded = Species('IF1_degraded')\n",
    "IF2 = Species('IF2')\n",
    "IF2_degraded = Species('IF2_degraded')\n",
    "IF2_GDP = Species('IF2_GDP')\n",
    "IF2_GTP = Species('IF2_GTP')\n",
    "IF2_GTP_fMettRNAfMetCAU = Species('IF2_GTP_fMettRNAfMetCAU')\n",
    "IF3 = Species('IF3')\n",
    "IF3_degraded = Species('IF3_degraded')\n",
    "MK = Species('MK')\n",
    "\n",
    "MK_ADP_1 = Species('MK_ADP_1')\n",
    "MK_ADP_2 = Species('MK_ADP_2')\n",
    "MK_ADP_ADP = Species('MK_ADP_ADP')\n",
    "MK_AMP = Species('MK_AMP')\n",
    "MK_ATP = Species('MK_ATP')\n",
    "MK_ATP_AMP = Species('MK_ATP_AMP')\n",
    "MK_degraded = Species('MK_degraded')\n",
    "mRNA_degraded = Species('mRNA_degraded')\n",
    "MTF = Species('MTF')\n",
    "\n",
    "MTF_degraded = Species('MTF_degraded')\n",
    "MTF_FD = Species('MTF_FD')\n",
    "MTF_THF = Species('MTF_THF')\n",
    "NDK = Species('NDK')\n",
    "NDK_ADP = Species('NDK_ADP')\n",
    "NDK_ATP = Species('NDK_ATP')\n",
    "NDK_degraded = Species('NDK_degraded')\n",
    "NDK_GDP = Species('NDK_GDP')\n",
    "NDK_GDP_ATP = Species('NDK_GDP_ATP')\n",
    "NDK_GTP = Species('NDK_GTP')\n",
    "\n",
    "NDK_GTP_ADP = Species('NDK_GTP_ADP')\n",
    "PPiase = Species('PPiase')\n",
    "PPiase_degraded = Species('PPiase_degraded')\n",
    "PPiase_PO4 = Species('PPiase_PO4')\n",
    "PPiase_PO4_PO4 = Species('PPiase_PO4_PO4')\n",
    "PPiase_PPi = Species('PPiase_PPi')\n",
    "RF1 = Species('RF1')\n",
    "RF1_degraded = Species('RF1_degraded')\n",
    "\n",
    "RF2 = Species('RF2')\n",
    "RF2_degraded = Species('RF2_degraded')\n",
    "RF3 = Species('RF3')\n",
    "RF3_degraded = Species('RF3_degraded')\n",
    "RF3_GDP = Species('RF3_GDP')\n",
    "RF3_GTP = Species('RF3_GTP')\n",
    "RRF = Species('RRF')\n",
    "RRF_degraded = Species('RRF_degraded')\n",
    "\n",
    "RS30S = Species('RS30S')\n",
    "RS30S_degraded = Species('RS30S_degraded')\n",
    "RS30S_IF1_IF3_IF2_GTP_mRNA = Species('RS30S_IF1_IF3_IF2_GTP_mRNA')\n",
    "RS30S_IF1_IF3_mRNA = Species('RS30S_IF1_IF3_mRNA')\n",
    "RS30S_IF1_mRNA = Species('RS30S_IF1_mRNA')\n",
    "RS30S_IF1 = Species('RS30S_IF1')\n",
    "RS30S_IF1_IF2_GTP = Species('RS30S_IF1_IF2_GTP')\n",
    "RS30S_IF1_IF2_GTP_mRNA = Species('RS30S_IF1_IF2_GTP_mRNA')\n",
    "RS30S_IF1_IF3 = Species('RS30S_IF1_IF3')\n",
    "RS30S_IF1_IF3_IF2_GTP = Species('RS30S_IF1_IF3_IF2_GTP')\n",
    "RS30S_IF3_IF2_GTP = Species('RS30S_IF3_IF2_GTP')\n",
    "\n",
    "RS30S_IF3_IF2_GTP_mRNA = Species('RS30S_IF3_IF2_GTP_mRNA')      \n",
    "RS30S_IF2_GTP = Species('RS30S_IF2_GTP')\n",
    "RS30S_IF2_GTP_mRNA = Species('RS30S_IF2_GTP_mRNA')\n",
    "RS30S_IF3 = Species('RS30S_IF3')\n",
    "RS30S_IF3_mRNA = Species('RS30S_IF3_mRNA')\n",
    "RS30S_mRNA = Species('RS30S_mRNA')\n",
    "RS50S = Species('RS50S')\n",
    "RS50S_degraded = Species('RS50S_degraded')\n",
    "RS50S_EFG_GDP = Species('RS50S_EFG_GDP')\n",
    "RS50S_EFG_GDP_PO4 = Species('RS50S_EFG_GDP_PO4')\n",
    "\n",
    "RS50S_EFG_GTP = Species('RS50S_EFG_GTP')\n",
    "RS50S_RRF = Species('RS50S_RRF')\n",
    "RS50S_RRF_EFG_GDP = Species('RS50S_RRF_EFG_GDP')\n",
    "RS70S = Species('RS70S')\n",
    "RS70S_EFG_GDP = Species('RS70S_EFG_GDP')\n",
    "RS70S_EFG_GDP_PO4 = Species('RS70S_EFG_GDP_PO4')\n",
    "RS70S_EFG_GTP = Species('RS70S_EFG_GTP')\n",
    "RS70S_IF1 = Species('RS70S_IF1')\n",
    "RS70S_IF3 = Species('RS70S_IF3')\n",
    "RS70S_IF1_IF3 = Species('RS70S_IF1_IF3')         \n",
    "\n",
    "THF = Species('THF')\n",
    "fMet = Species('fMet')\n",
    "fMet_degraded = Species('fMet_degraded')\n",
    "\n",
    "#######################################################################################################################\n",
    "\n",
    "#Creates a list of all the general species\n",
    "list_species_gen=[ADP, AMP, CK, CK_ADP, CK_ATP, \n",
    "                  CK_CP, CK_CP_ADP, CK_Cr, CK_Cr_ATP, CK_degraded,\n",
    "                  CP, Cr, EFG, EFG_degraded, EFG_GDP,\n",
    "                  EFG_GTP, EFTs, EFTs_degraded, EFTu, EFTu_degraded,\n",
    "                  EFTu_EFTs, EFTu_GDP, EFTu_GDP_EFTs, EFTu_GTP, EFTu_GTP_EFTs,\n",
    "                  FD,  GMP, IF1, #PO4, PPi, ATP, GTP,GDP,\n",
    "                  IF1_degraded, IF2, IF2_degraded, IF2_GDP, IF2_GTP,\n",
    "                  IF2_GTP_fMettRNAfMetCAU, IF3, IF3_degraded, MK, MK_ADP_1, MK_ADP_2,\n",
    "                  MK_ADP_ADP, MK_AMP, MK_ATP, MK_ATP_AMP, MK_degraded,\n",
    "                  mRNA, mRNA_degraded, MTF, MTF_degraded, MTF_FD, MTF_THF,\n",
    "                  NDK, NDK_ADP, NDK_ATP, NDK_degraded, NDK_GDP, NDK_GDP_ATP,\n",
    "                  NDK_GTP, NDK_GTP_ADP, PPiase, PPiase_degraded,\n",
    "                  PPiase_PO4, PPiase_PO4_PO4, PPiase_PPi, RF1, RF1_degraded,\n",
    "                  RF2, RF2_degraded, RF3, RF3_degraded, RF3_GDP,\n",
    "                  RF3_GTP, RRF, RRF_degraded, RS30S, RS30S_degraded,\n",
    "                  RS30S_IF1_IF3_IF2_GTP_mRNA, RS30S_IF1_IF3_mRNA, RS30S_IF1_mRNA, RS30S_IF1, RS30S_IF1_IF2_GTP,\n",
    "                  RS30S_IF1_IF2_GTP_mRNA, RS30S_IF1_IF3, RS30S_IF1_IF3_IF2_GTP, RS30S_IF3_IF2_GTP_mRNA,   \n",
    "                  RS30S_IF2_GTP, RS30S_IF2_GTP_mRNA, RS30S_IF3, RS30S_IF3_mRNA, RS30S_IF3_IF2_GTP,\n",
    "                  RS30S_mRNA, RS50S, RS50S_degraded, RS50S_EFG_GDP, RS50S_EFG_GDP_PO4,\n",
    "                  RS50S_EFG_GTP, RS50S_RRF, RS50S_RRF_EFG_GDP, RS70S, RS70S_EFG_GDP,\n",
    "                  RS70S_EFG_GDP_PO4, RS70S_EFG_GTP, RS70S_IF1, RS70S_IF3, RS70S_IF1_IF3,        \n",
    "                  THF, fMet, fMet_degraded,]\n",
    "\n",
    "#######################################################################################################################\n",
    "\n",
    "#Creates a list of all the reactions only involving the general species. The reaction rates are read from the rxn_k dictionary created initially.\n",
    "list_of_reaction_gen= [    \n",
    "    Reaction.from_massaction([EFTu,EFTs],[EFTu_EFTs], k_forward = rxn_k['re0000000261_k1']),\n",
    "    Reaction.from_massaction([EFTu_EFTs],[EFTu,EFTs], k_forward = rxn_k['re0000000262_k1']),\n",
    "    Reaction.from_massaction([EFTu_EFTs,GDP],[EFTu_GDP_EFTs], k_forward = rxn_k['re0000000263_k1']),\n",
    "    Reaction.from_massaction([EFTu_GDP_EFTs],[EFTu_EFTs,GDP], k_forward = rxn_k['re0000000264_k1']),\n",
    "    Reaction.from_massaction([EFTu_GDP_EFTs],[EFTu_GDP,EFTs], k_forward = rxn_k['re0000000265_k1']),\n",
    "    Reaction.from_massaction([EFTu_GDP,EFTs],[EFTu_GDP_EFTs], k_forward = rxn_k['re0000000266_k1']),\n",
    "    Reaction.from_massaction([EFTu_GDP],[EFTu,GDP], k_forward = rxn_k['re0000000267_k1']),\n",
    "    Reaction.from_massaction([EFTu,GDP],[EFTu_GDP], k_forward = rxn_k['re0000000268_k1']),\n",
    "    Reaction.from_massaction([EFTu_EFTs,GTP],[EFTu_GTP_EFTs], k_forward = rxn_k['re0000000269_k1']),\n",
    "    Reaction.from_massaction([EFTu_GTP_EFTs],[EFTu_EFTs,GTP], k_forward = rxn_k['re0000000270_k1']),\n",
    "    Reaction.from_massaction([EFTu_GTP_EFTs],[EFTu_GTP,EFTs], k_forward = rxn_k['re0000000271_k1']),\n",
    "    Reaction.from_massaction([EFTu_GTP,EFTs],[EFTu_GTP_EFTs], k_forward = rxn_k['re0000000272_k1']),\n",
    "    Reaction.from_massaction([EFTu,GTP],[EFTu_GTP], k_forward = rxn_k['re0000000273_k1']),\n",
    "    Reaction.from_massaction([EFTu_GTP],[EFTu,GTP], k_forward = rxn_k['re0000000274_k1']),\n",
    "\n",
    "    Reaction.from_massaction([EFG_GDP],[EFG,GDP], k_forward = rxn_k['re0000000292_k1']),\n",
    "    Reaction.from_massaction([EFG,GDP],[EFG_GDP], k_forward = rxn_k['re0000000293_k1']),\n",
    "    Reaction.from_massaction([EFG,GTP],[EFG_GTP], k_forward = rxn_k['re0000000294_k1']),\n",
    "    Reaction.from_massaction([EFG_GTP],[EFG,GTP], k_forward = rxn_k['re0000000295_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS50S,EFG_GTP],[RS50S_EFG_GTP], k_forward = rxn_k['re0000000298_k1']),\n",
    "    Reaction.from_massaction([RS50S_EFG_GTP],[RS50S,EFG_GTP], k_forward = rxn_k['re0000000299_k1']),\n",
    "    Reaction.from_massaction([RS70S,EFG_GTP],[RS70S_EFG_GTP], k_forward = rxn_k['re0000000300_k1']),\n",
    "    Reaction.from_massaction([RS70S_EFG_GTP],[RS70S,EFG_GTP], k_forward = rxn_k['re0000000301_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS50S_EFG_GTP],[RS50S_EFG_GDP_PO4], k_forward = rxn_k['re0000000302_k1']),\n",
    "    Reaction.from_massaction([RS50S_EFG_GDP_PO4],[RS50S_EFG_GTP], k_forward = rxn_k['re0000000303_k1']),\n",
    "    Reaction.from_massaction([RS70S_EFG_GTP],[RS70S_EFG_GDP_PO4], k_forward = rxn_k['re0000000304_k1']),\n",
    "    Reaction.from_massaction([RS70S_EFG_GDP_PO4],[RS70S_EFG_GTP], k_forward = rxn_k['re0000000305_k1']),\n",
    "    Reaction.from_massaction([RS50S_EFG_GDP_PO4],[RS50S_EFG_GDP,PO4,], k_forward = rxn_k['re0000000306_k1']),\n",
    "    Reaction.from_massaction([RS70S_EFG_GDP_PO4],[RS70S_EFG_GDP,PO4,], k_forward = rxn_k['re0000000307_k1']),\n",
    "    Reaction.from_massaction([RS50S_EFG_GDP],[RS50S,EFG_GDP], k_forward = rxn_k['re0000000308_k1']),\n",
    "    Reaction.from_massaction([RS70S_EFG_GDP],[RS70S,EFG_GDP], k_forward = rxn_k['re0000000309_k1']),\n",
    "\n",
    "    Reaction.from_massaction([CK,ADP],[CK_ADP], k_forward = rxn_k['re0000000330_k1']),\n",
    "    Reaction.from_massaction([CK_ADP],[CK,ADP], k_forward = rxn_k['re0000000331_k1']),\n",
    "\n",
    "    Reaction.from_massaction([CK,CP],[CK_CP], k_forward = rxn_k['re0000000332_k1']),\n",
    "    Reaction.from_massaction([CK_CP],[CK,CP], k_forward = rxn_k['re0000000333_k1']),\n",
    "    Reaction.from_massaction([CK_CP,ADP],[CK_CP_ADP], k_forward = rxn_k['re0000000334_k1']),\n",
    "    Reaction.from_massaction([CK_CP_ADP],[CK_CP,ADP], k_forward = rxn_k['re0000000335_k1']),\n",
    "    Reaction.from_massaction([CK_ADP,CP],[CK_CP_ADP], k_forward = rxn_k['re0000000336_k1']),\n",
    "    Reaction.from_massaction([CK_CP_ADP],[CP,CK_ADP], k_forward = rxn_k['re0000000337_k1']),\n",
    "    Reaction.from_massaction([CK_CP_ADP],[CK_Cr_ATP], k_forward = rxn_k['re0000000338_k1']),\n",
    "    Reaction.from_massaction([CK_Cr_ATP],[CK_CP_ADP], k_forward = rxn_k['re0000000339_k1']),\n",
    "    Reaction.from_massaction([CK_Cr_ATP],[CK_Cr,ATP], k_forward = rxn_k['re0000000340_k1']),\n",
    "    Reaction.from_massaction([CK_Cr,ATP],[CK_Cr_ATP], k_forward = rxn_k['re0000000341_k1']),\n",
    "\n",
    "    Reaction.from_massaction([CK_Cr_ATP],[CK_ATP,Cr], k_forward = rxn_k['re0000000342_k1']),\n",
    "    Reaction.from_massaction([CK_ATP,Cr],[CK_Cr_ATP], k_forward = rxn_k['re0000000343_k1']),\n",
    "    Reaction.from_massaction([CK_ATP],[CK,ATP], k_forward = rxn_k['re0000000344_k1']),\n",
    "    Reaction.from_massaction([CK,ATP],[CK_ATP], k_forward = rxn_k['re0000000345_k1']),\n",
    "    Reaction.from_massaction([CK_Cr],[CK,Cr], k_forward = rxn_k['re0000000346_k1']),\n",
    "    Reaction.from_massaction([CK,Cr],[CK_Cr], k_forward = rxn_k['re0000000347_k1']),\n",
    "\n",
    "    Reaction.from_massaction([NDK,ATP],[NDK_ATP], k_forward = rxn_k['re0000000355_k1']),\n",
    "    Reaction.from_massaction([NDK_ATP],[NDK,ATP], k_forward = rxn_k['re0000000356_k1']),\n",
    "    Reaction.from_massaction([NDK,GDP],[NDK_GDP], k_forward = rxn_k['re0000000357_k1']),\n",
    "    Reaction.from_massaction([NDK_GDP],[NDK,GDP], k_forward = rxn_k['re0000000358_k1']),\n",
    "    Reaction.from_massaction([NDK_GDP,ATP],[NDK_GDP_ATP], k_forward = rxn_k['re0000000359_k1']),\n",
    "    Reaction.from_massaction([NDK_GDP_ATP],[NDK_GDP,ATP], k_forward = rxn_k['re0000000360_k1']),\n",
    "    Reaction.from_massaction([NDK_ATP,GDP],[NDK_GDP_ATP], k_forward = rxn_k['re0000000361_k1']),\n",
    "\n",
    "    Reaction.from_massaction([NDK_GDP_ATP],[NDK_ATP,GDP], k_forward = rxn_k['re0000000362_k1']),\n",
    "    Reaction.from_massaction([NDK_GDP_ATP],[NDK_GTP_ADP], k_forward = rxn_k['re0000000363_k1']),\n",
    "\n",
    "    Reaction.from_massaction([NDK_GTP_ADP],[NDK_ADP,GTP], k_forward = rxn_k['re0000000365_k1']),\n",
    "    Reaction.from_massaction([NDK_GTP_ADP],[NDK_GTP,ADP], k_forward = rxn_k['re0000000366_k1']),\n",
    "    Reaction.from_massaction([NDK_ADP],[NDK,ADP], k_forward = rxn_k['re0000000367_k1']),\n",
    "    Reaction.from_massaction([NDK_GTP],[NDK,GTP], k_forward = rxn_k['re0000000368_k1']),\n",
    "\n",
    "    Reaction.from_massaction([NDK_GTP,ADP],[NDK_GTP_ADP], k_forward = rxn_k['re0000000375_k1']),\n",
    "    Reaction.from_massaction([NDK_ADP,GTP],[NDK_GTP_ADP], k_forward = rxn_k['re0000000376_k1']),\n",
    "    Reaction.from_massaction([NDK,GTP],[NDK_GTP], k_forward = rxn_k['re0000000377_k1']),\n",
    "    Reaction.from_massaction([NDK,ADP],[NDK_ADP], k_forward = rxn_k['re0000000378_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MK,ATP],[MK_ATP], k_forward = rxn_k['re0000000380_k1']),\n",
    "    Reaction.from_massaction([MK_ATP],[ATP,MK], k_forward = rxn_k['re0000000381_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MK,AMP],[MK_AMP], k_forward = rxn_k['re0000000382_k1']),\n",
    "    Reaction.from_massaction([MK_AMP],[MK,AMP], k_forward = rxn_k['re0000000383_k1']),\n",
    "    Reaction.from_massaction([MK_AMP,ATP],[MK_ATP_AMP], k_forward = rxn_k['re0000000384_k1']),\n",
    "    Reaction.from_massaction([MK_ATP_AMP],[ATP,MK_AMP], k_forward = rxn_k['re0000000385_k1']),\n",
    "    Reaction.from_massaction([MK_ATP,AMP],[MK_ATP_AMP], k_forward = rxn_k['re0000000386_k1']),\n",
    "    Reaction.from_massaction([MK_ATP_AMP],[MK_ATP,AMP], k_forward = rxn_k['re0000000387_k1']),\n",
    "    Reaction.from_massaction([MK_ATP_AMP],[MK_ADP_ADP], k_forward = rxn_k['re0000000388_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_ADP],[MK_ATP_AMP], k_forward = rxn_k['re0000000389_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_ADP],[MK_ADP_1,ADP], k_forward = rxn_k['re0000000390_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_1,ADP],[MK_ADP_ADP], k_forward = rxn_k['re0000000391_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MK_ADP_ADP],[MK_ADP_2,ADP], k_forward = rxn_k['re0000000392_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_2,ADP],[MK_ADP_ADP], k_forward = rxn_k['re0000000393_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_1],[MK,ADP], k_forward = rxn_k['re0000000394_k1']),\n",
    "    Reaction.from_massaction([MK,ADP],[MK_ADP_1], k_forward = rxn_k['re0000000395_k1']),\n",
    "    Reaction.from_massaction([MK_ADP_2],[ADP,MK], k_forward = rxn_k['re0000000396_k1']),\n",
    "    Reaction.from_massaction([MK,ADP],[MK_ADP_2], k_forward = rxn_k['re0000000397_k1']),\n",
    "\n",
    "    Reaction.from_massaction([PPiase,PPi],[PPiase_PPi], k_forward = rxn_k['re0000000405_k1']),\n",
    "    Reaction.from_massaction([PPiase_PPi],[PPiase,PPi], k_forward = rxn_k['re0000000406_k1']),\n",
    "    Reaction.from_massaction([PPiase_PPi],[PPiase_PO4_PO4], k_forward = rxn_k['re0000000407_k1']),\n",
    "    Reaction.from_massaction([PPiase_PO4_PO4],[PPiase_PPi], k_forward = rxn_k['re0000000408_k1']),\n",
    "    Reaction.from_massaction([PPiase_PO4_PO4],[PPiase_PO4,PO4], k_forward = rxn_k['re0000000409_k1']),\n",
    "    Reaction.from_massaction([PPiase_PO4,PO4],[PPiase_PO4_PO4], k_forward = rxn_k['re0000000410_k1']),\n",
    "    Reaction.from_massaction([PPiase_PO4],[PPiase,PO4], k_forward = rxn_k['re0000000411_k1']),\n",
    "\n",
    "    Reaction.from_massaction([PPiase,PO4],[PPiase_PO4], k_forward = rxn_k['re0000000412_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF,FD],[MTF_FD], k_forward = rxn_k['re0000000418_k1']),\n",
    "    Reaction.from_massaction([MTF_FD],[MTF,FD], k_forward = rxn_k['re0000000419_k1']),\n",
    "    Reaction.from_massaction([MTF_THF],[MTF,THF], k_forward = rxn_k['re0000000434_k1']),\n",
    "\n",
    "    Reaction.from_massaction([IF2,GTP],[IF2_GTP], k_forward = rxn_k['re0000000445_k1']),\n",
    "    Reaction.from_massaction([IF2_GTP],[GTP,IF2], k_forward = rxn_k['re0000000446_k1']),\n",
    "    Reaction.from_massaction([IF2,GDP],[IF2_GDP], k_forward = rxn_k['re0000000447_k1']),\n",
    "    Reaction.from_massaction([IF2_GDP],[IF2,GDP], k_forward = rxn_k['re0000000448_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S],[RS30S,RS50S], k_forward = rxn_k['re0000000455_k1']),\n",
    "    Reaction.from_massaction([RS30S,RS50S],[RS70S], k_forward = rxn_k['re0000000456_k1']),\n",
    "    Reaction.from_massaction([RS70S,IF3],[RS70S_IF3], k_forward = rxn_k['re0000000457_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3],[RS70S,IF3], k_forward = rxn_k['re0000000458_k1']),\n",
    "    Reaction.from_massaction([RS30S,IF3],[RS30S_IF3], k_forward = rxn_k['re0000000459_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3],[RS30S,IF3], k_forward = rxn_k['re0000000460_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3],[RS30S_IF3,RS50S], k_forward = rxn_k['re0000000461_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3,RS50S],[RS70S_IF3], k_forward = rxn_k['re0000000462_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3,IF2_GTP],[RS30S_IF3_IF2_GTP], k_forward = rxn_k['re0000000463_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP],[RS30S_IF3,IF2_GTP], k_forward = rxn_k['re0000000464_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF3,mRNA],[RS30S_IF3_mRNA], k_forward= rxn_k['re0000000469_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA],[RS30S_IF3,mRNA], k_forward= rxn_k['re0000000470_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA,IF2_GTP],[RS30S_IF3_IF2_GTP_mRNA], k_forward= rxn_k['re0000000471_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_mRNA],[RS30S_IF3_mRNA,IF2_GTP], k_forward= rxn_k['re0000000472_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP,mRNA],[RS30S_IF3_IF2_GTP_mRNA], k_forward= rxn_k['re0000000481_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_mRNA],[RS30S_IF3_IF2_GTP,mRNA], k_forward= rxn_k['re0000000482_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1],[RS30S_IF1,RS50S], k_forward = rxn_k['re0000000487_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1,RS50S],[RS70S_IF1], k_forward = rxn_k['re0000000488_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1,IF3],[RS70S_IF1_IF3], k_forward = rxn_k['re0000000489_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3],[RS70S_IF1,IF3], k_forward = rxn_k['re0000000490_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1,IF3],[RS30S_IF1_IF3], k_forward = rxn_k['re0000000491_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3],[RS30S_IF1,IF3], k_forward = rxn_k['re0000000492_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3],[RS30S_IF1_IF3,RS50S], k_forward = rxn_k['re0000000493_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3,RS50S],[RS70S_IF1_IF3], k_forward = rxn_k['re0000000494_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3,IF2_GTP],[RS30S_IF1_IF3_IF2_GTP], k_forward = rxn_k['re0000000495_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP],[RS30S_IF1_IF3,IF2_GTP], k_forward = rxn_k['re0000000496_k1']),\n",
    "    Reaction.from_massaction([RS70S,IF1],[RS70S_IF1], k_forward= rxn_k['re0000000501_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1],[RS70S,IF1], k_forward = rxn_k['re0000000502_k1']),\n",
    "    Reaction.from_massaction([RS30S,IF1],[RS30S_IF1], k_forward = rxn_k['re0000000503_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1],[RS30S,IF1], k_forward = rxn_k['re0000000504_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF3,IF1],[RS70S_IF1_IF3], k_forward = rxn_k['re0000000505_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3],[RS70S_IF3,IF1], k_forward = rxn_k['re0000000506_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3,IF1],[RS30S_IF1_IF3], k_forward= rxn_k['re0000000507_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3],[RS30S_IF3,IF1], k_forward = rxn_k['re0000000508_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP,IF1],[RS30S_IF1_IF3_IF2_GTP], k_forward= rxn_k['re0000000509_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP],[RS30S_IF3_IF2_GTP,IF1], k_forward = rxn_k['re0000000510_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3,mRNA],[RS30S_IF1_IF3_mRNA], k_forward = rxn_k['re0000000513_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA],[RS30S_IF1_IF3,mRNA], k_forward = rxn_k['re0000000514_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA,IF2_GTP],[RS30S_IF1_IF3_IF2_GTP_mRNA], k_forward = rxn_k['re0000000515_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_mRNA],[RS30S_IF1_IF3_mRNA,IF2_GTP], k_forward = rxn_k['re0000000516_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP,mRNA],[RS30S_IF1_IF3_IF2_GTP_mRNA], k_forward = rxn_k['re0000000525_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_mRNA],[RS30S_IF1_IF3_IF2_GTP,mRNA], k_forward = rxn_k['re0000000526_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA,IF1],[RS30S_IF1_IF3_mRNA], k_forward = rxn_k['re0000000531_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA],[RS30S_IF3_mRNA,IF1], k_forward = rxn_k['re0000000532_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_mRNA,IF1],[RS30S_IF1_IF3_IF2_GTP_mRNA], k_forward= rxn_k['re0000000535_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_mRNA],[RS30S_IF3_IF2_GTP_mRNA,IF1], k_forward = rxn_k['re0000000536_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP],[IF2_GTP,RS30S], k_forward= rxn_k['re0000000610_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP,IF3],[RS30S_IF3_IF2_GTP], k_forward= rxn_k['re0000000615_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP],[IF3,RS30S_IF2_GTP], k_forward= rxn_k['re0000000616_k1']),\n",
    "    Reaction.from_massaction([mRNA,RS30S],[RS30S_mRNA], k_forward = rxn_k['re0000000619_k1']),\n",
    "    Reaction.from_massaction([RS30S_mRNA],[RS30S,mRNA], k_forward= rxn_k['re0000000620_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_mRNA,IF2_GTP],[RS30S_IF2_GTP_mRNA], k_forward= rxn_k['re0000000625_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_mRNA],[IF2_GTP,RS30S_mRNA], k_forward= rxn_k['re0000000626_k1']),\n",
    "    Reaction.from_massaction([mRNA,RS30S_IF2_GTP],[RS30S_IF2_GTP_mRNA], k_forward = rxn_k['re0000000631_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_mRNA],[RS30S_IF2_GTP,mRNA], k_forward= rxn_k['re0000000632_k1']),\n",
    "    Reaction.from_massaction([RS30S_mRNA,IF3],[RS30S_IF3_mRNA], k_forward= rxn_k['re0000000635_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA],[IF3,RS30S_mRNA], k_forward= rxn_k['re0000000636_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_mRNA,IF3],[RS30S_IF3_IF2_GTP_mRNA], k_forward= rxn_k['re0000000639_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_mRNA],[IF3,RS30S_IF2_GTP_mRNA], k_forward= rxn_k['re0000000640_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1,IF2_GTP],[RS30S_IF1_IF2_GTP], k_forward = rxn_k['re0000000643_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP],[IF2_GTP,RS30S_IF1], k_forward = rxn_k['re0000000644_k1']),\n",
    "\n",
    "    Reaction.from_massaction([IF1,RS30S_IF2_GTP],[RS30S_IF1_IF2_GTP], k_forward = rxn_k['re0000000649_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP],[IF1,RS30S_IF2_GTP], k_forward = rxn_k['re0000000650_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP,IF3],[RS30S_IF1_IF3_IF2_GTP], k_forward = rxn_k['re0000000653_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP],[IF3,RS30S_IF1_IF2_GTP], k_forward = rxn_k['re0000000654_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA,IF2_GTP],[RS30S_IF1_IF2_GTP_mRNA], k_forward= rxn_k['re0000000657_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_mRNA],[RS30S_IF1_mRNA,IF2_GTP], k_forward = rxn_k['re0000000658_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA,IF3],[RS30S_IF1_IF3_mRNA], k_forward= rxn_k['re0000000667_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA],[IF3,RS30S_IF1_mRNA], k_forward = rxn_k['re0000000668_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_mRNA,IF3],[RS30S_IF1_IF3_IF2_GTP_mRNA], k_forward = rxn_k['re0000000671_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_mRNA],[IF3,RS30S_IF1_IF2_GTP_mRNA], k_forward = rxn_k['re0000000672_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1,mRNA],[RS30S_IF1_mRNA], k_forward = rxn_k['re0000000675_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA],[RS30S_IF1,mRNA], k_forward= rxn_k['re0000000676_k1']),\n",
    "    Reaction.from_massaction([RS30S_mRNA,IF1],[RS30S_IF1_mRNA], k_forward= rxn_k['re0000000677_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA],[RS30S_mRNA,IF1], k_forward= rxn_k['re0000000678_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_mRNA,IF1],[RS30S_IF1_IF2_GTP_mRNA], k_forward= rxn_k['re0000000679_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_mRNA],[RS30S_IF2_GTP_mRNA,IF1], k_forward = rxn_k['re0000000680_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP,mRNA],[RS30S_IF1_IF2_GTP_mRNA], k_forward = rxn_k['re0000000681_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_mRNA],[RS30S_IF1_IF2_GTP,mRNA], k_forward = rxn_k['re0000000682_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RF3_GDP],[GDP,RF3], k_forward = rxn_k['re0000000825_k1']),\n",
    "    Reaction.from_massaction([RF3,GDP],[RF3_GDP], k_forward = rxn_k['re0000000826_k1']),\n",
    "    Reaction.from_massaction([RF3,GTP],[RF3_GTP], k_forward = rxn_k['re0000000827_k1']),\n",
    "    Reaction.from_massaction([RF3_GTP],[RF3,GTP], k_forward = rxn_k['re0000000828_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS50S_RRF],[RS50S,RRF], k_forward= rxn_k['re0000000918_k1']),\n",
    "    Reaction.from_massaction([RS50S_RRF_EFG_GDP],[RS50S_RRF,EFG_GDP], k_forward= rxn_k['re0000000922_k1']),\n",
    "    Reaction.from_massaction([RS50S_RRF_EFG_GDP],[RS50S_EFG_GDP,RRF], k_forward= rxn_k['re0000000923_k1']),\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f136249d-3b8b-4b6b-a400-aeff178ef659",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Met Species and Reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "353190e1-9fc1-4a46-8afd-8a63975e5a0b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Create al the fMet and needed MetRS Species\n",
    "fMettRNAfMetCAU = Species('fMettRNAfMetCAU')\n",
    "fMettRNAfMetCAU_degraded = Species('fMettRNAfMetCAU_degraded')\n",
    "tRNAfMetCAU = Species('tRNAfMetCAU')\n",
    "tRNAfMetCAU_degraded = Species('tRNAfMetCAU_degraded')\n",
    "EFTu_GTP_MettRNAfMetCAU = Species('EFTu_GTP_MettRNAfMetCAU')\n",
    "\n",
    "Met = Species('Met')\n",
    "MetAMP = Species('MetAMP')\n",
    "MetRS = Species('MetRS')\n",
    "\n",
    "MetRS_AMP = Species('MetRS_AMP')\n",
    "MetRS_AMP_MettRNAfMetCAU = Species('MetRS_AMP_MettRNAfMetCAU')\n",
    "MetRS_ATP = Species('MetRS_ATP')\n",
    "MetRS_ATP_tRNAfMetCAU = Species('MetRS_ATP_tRNAfMetCAU')\n",
    "MetRS_degraded = Species('MetRS_degraded')\n",
    "MetRS_Met = Species('MetRS_Met')\n",
    "MetRS_Met_ATP = Species('MetRS_Met_ATP')\n",
    "MetRS_Met_ATP_tRNAfMetCAU = Species('MetRS_Met_ATP_tRNAfMetCAU')\n",
    "MetRS_Met_tRNAfMetCAU = Species('MetRS_Met_tRNAfMetCAU')\n",
    "MetRS_MetAMP = Species('MetRS_MetAMP')\n",
    "MetRS_MetAMP_PPi = Species('MetRS_MetAMP_PPi')\n",
    "MetRS_MetAMP_PPi_tRNAfMetCAU = Species('MetRS_MetAMP_PPi_tRNAfMetCAU')\n",
    "MetRS_MetAMP_tRNAfMetCAU = Species('MetRS_MetAMP_tRNAfMetCAU')\n",
    "MetRS_MettRNAfMetCAU=Species('MetRS_MettRNAfMetCAU')\n",
    "MetRS_tRNAfMetCAU = Species('MetRS_tRNAfMetCAU')\n",
    "\n",
    "MettRNAfMetCAU = Species('MettRNAfMetCAU')\n",
    "MettRNAfMetCAU_degraded = Species('MettRNAfMetCAU_degraded')\n",
    "MTF_FD_MettRNAfMetCAU = Species('MTF_FD_MettRNAfMetCAU')\n",
    "MTF_fMettRNAfMetCAU = Species('MTF_fMettRNAfMetCAU')\n",
    "MTF_MettRNAfMetCAU = Species('MTF_MettRNAfMetCAU')\n",
    "MTF_THF_fMettRNAfMetCAU = Species('MTF_THF_fMettRNAfMetCAU')\n",
    "\n",
    "RS30S_fMettRNAfMetCAU_mRNA = Species('RS30S_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF1_fMettRNAfMetCAU_mRNA = Species('RS30S_IF1_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF1_IF2_GTP_fMettRNAfMetCAU = Species('RS30S_IF1_IF2_GTP_fMettRNAfMetCAU')\n",
    "RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA = Species('RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU = Species('RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU')\n",
    "RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF2_GTP_fMettRNAfMetCAU = Species('RS30S_IF2_GTP_fMettRNAfMetCAU')\n",
    "RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF3_fMettRNAfMetCAU_mRNA = Species('RS30S_IF3_fMettRNAfMetCAU_mRNA')\n",
    "RS30S_IF3_IF2_GTP_fMettRNAfMetCAU = Species('RS30S_IF3_IF2_GTP_fMettRNAfMetCAU')\n",
    "RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "\n",
    "RS70S_IF1_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF2_GDP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF2_GDP_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF3_fMettRNAfMetCAU_mRNA = Species('RS70S_IF3_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA = Species('RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA')\n",
    "RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA = Species('RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA')\n",
    "\n",
    "#######################################################################################################################\n",
    "\n",
    "#A list of Species in relation to fMet\n",
    "list_species_fmet=[\n",
    "    fMettRNAfMetCAU, fMettRNAfMetCAU_degraded, Met, MetAMP, MetRS,tRNAfMetCAU, tRNAfMetCAU_degraded, EFTu_GTP_MettRNAfMetCAU,\n",
    "    \n",
    "    MetRS_AMP, MetRS_AMP_MettRNAfMetCAU, MetRS_ATP, MetRS_ATP_tRNAfMetCAU, MetRS_degraded,\n",
    "    MetRS_Met, MetRS_Met_ATP, MetRS_Met_ATP_tRNAfMetCAU, MetRS_Met_tRNAfMetCAU, MetRS_MetAMP,\n",
    "    MetRS_MetAMP_PPi, MetRS_MetAMP_PPi_tRNAfMetCAU, MetRS_MetAMP_tRNAfMetCAU, MetRS_MettRNAfMetCAU, MetRS_tRNAfMetCAU,\n",
    "    MettRNAfMetCAU, MettRNAfMetCAU_degraded, MTF_FD_MettRNAfMetCAU, MTF_fMettRNAfMetCAU, MTF_MettRNAfMetCAU, MTF_THF_fMettRNAfMetCAU,\n",
    "    \n",
    "    RS30S_fMettRNAfMetCAU_mRNA, RS30S_IF1_fMettRNAfMetCAU_mRNA, RS30S_IF1_IF2_GTP_fMettRNAfMetCAU, \n",
    "    RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA, RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA, RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU, \n",
    "    RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA, RS30S_IF2_GTP_fMettRNAfMetCAU, RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA, \n",
    "    RS30S_IF3_fMettRNAfMetCAU_mRNA, RS30S_IF3_IF2_GTP_fMettRNAfMetCAU, RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,\n",
    "\n",
    "    RS70S_IF1_fMettRNAfMetCAU_mRNA, RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA, RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA,\n",
    "    RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA, RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA, RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,\n",
    "    RS70S_IF2_GDP_fMettRNAfMetCAU_mRNA, RS70S_IF3_fMettRNAfMetCAU_mRNA, RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA,\n",
    "    RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA, RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA]\n",
    "\n",
    "#######################################################################################################################\n",
    "\n",
    "#fMet Reactions \n",
    "list_reaction_fmet=[\n",
    "    Reaction.from_massaction([MetRS,Met],[MetRS_Met], k_forward= rxn_k['re0000000151_k1']),\n",
    "    Reaction.from_massaction([MetRS_MetAMP_PPi],[MetRS_MetAMP,PPi,], k_forward= rxn_k['re0000000152_k1']),#Mg\n",
    "\n",
    "    Reaction.from_massaction([MetRS_Met],[MetRS,Met], k_forward= rxn_k['re0000000156_k1']),\n",
    "    Reaction.from_massaction([MetRS,ATP],[MetRS_ATP], k_forward= rxn_k['re0000000157_k1']),\n",
    "    Reaction.from_massaction([MetRS_ATP],[MetRS,ATP], k_forward= rxn_k['re0000000158_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_ATP,Met],[MetRS_Met_ATP], k_forward= rxn_k['re0000000159_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP],[MetRS_ATP,Met], k_forward= rxn_k['re0000000160_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met,ATP],[MetRS_Met_ATP], k_forward= rxn_k['re0000000161_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP],[MetRS_Met,ATP], k_forward= rxn_k['re0000000162_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_Met_ATP],[MetRS_MetAMP_PPi], k_forward= rxn_k['re0000000165_k1']),\n",
    "    Reaction.from_massaction([MetRS_MetAMP_PPi],[MetRS_Met_ATP], k_forward= rxn_k['re0000000166_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_AMP],[MetRS,AMP], k_forward= rxn_k['re0000000170_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_MetAMP],[MetRS,MetAMP], k_forward= rxn_k['re0000000172_k1']),\n",
    "    Reaction.from_massaction([MetRS,MetAMP],[MetRS_MetAMP], k_forward= rxn_k['re0000000173_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_MetAMP_tRNAfMetCAU],[MetRS_AMP_MettRNAfMetCAU], k_forward= rxn_k['re0000000220_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_AMP_MettRNAfMetCAU],[MetRS_MettRNAfMetCAU,AMP], k_forward= rxn_k['re0000000222_k1']),\n",
    "    \n",
    "    Reaction.from_massaction([MetRS_AMP_MettRNAfMetCAU],[MettRNAfMetCAU,MetRS_AMP], k_forward= rxn_k['re0000000224_k1']),\n",
    "    Reaction.from_massaction([MetRS_AMP,MettRNAfMetCAU],[MetRS_AMP_MettRNAfMetCAU], k_forward= rxn_k['re0000000225_k1']),\n",
    "    Reaction.from_massaction([MetRS_MettRNAfMetCAU],[MetRS,MettRNAfMetCAU], k_forward= rxn_k['re0000000226_k1']),\n",
    "    Reaction.from_massaction([MetRS,MettRNAfMetCAU],[MetRS_MettRNAfMetCAU], k_forward= rxn_k['re0000000227_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_tRNAfMetCAU,Met],[MetRS_Met_tRNAfMetCAU], k_forward= rxn_k['re0000000230_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_MetAMP_PPi_tRNAfMetCAU],[MetRS_MetAMP_tRNAfMetCAU,PPi], k_forward= rxn_k['re0000000231_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_tRNAfMetCAU],[MetRS_tRNAfMetCAU,Met], k_forward= rxn_k['re0000000232_k1']),\n",
    "    Reaction.from_massaction([MetRS_tRNAfMetCAU,ATP],[MetRS_ATP_tRNAfMetCAU], k_forward= rxn_k['re0000000233_k1']),\n",
    "    Reaction.from_massaction([MetRS_ATP_tRNAfMetCAU],[MetRS_tRNAfMetCAU,ATP], k_forward= rxn_k['re0000000234_k1']),\n",
    "    Reaction.from_massaction([MetRS_ATP_tRNAfMetCAU,Met],[MetRS_Met_ATP_tRNAfMetCAU], k_forward= rxn_k['re0000000235_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP_tRNAfMetCAU],[MetRS_ATP_tRNAfMetCAU,Met], k_forward= rxn_k['re0000000236_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_tRNAfMetCAU,ATP],[MetRS_Met_ATP_tRNAfMetCAU], k_forward= rxn_k['re0000000237_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP_tRNAfMetCAU],[MetRS_Met_tRNAfMetCAU,ATP], k_forward= rxn_k['re0000000238_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP_tRNAfMetCAU],[MetRS_MetAMP_PPi_tRNAfMetCAU], k_forward= rxn_k['re0000000239_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS,tRNAfMetCAU],[MetRS_tRNAfMetCAU], k_forward= rxn_k['re0000000241_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met,tRNAfMetCAU],[MetRS_Met_tRNAfMetCAU], k_forward= rxn_k['re0000000242_k1']),\n",
    "    Reaction.from_massaction([MetRS_tRNAfMetCAU],[MetRS,tRNAfMetCAU], k_forward= rxn_k['re0000000243_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_tRNAfMetCAU],[MetRS_Met,tRNAfMetCAU], k_forward= rxn_k['re0000000244_k1']),\n",
    "    Reaction.from_massaction([MetRS_ATP,tRNAfMetCAU],[MetRS_ATP_tRNAfMetCAU], k_forward= rxn_k['re0000000245_k1']),\n",
    "    Reaction.from_massaction([MetRS_ATP_tRNAfMetCAU],[MetRS_ATP,tRNAfMetCAU], k_forward= rxn_k['re0000000246_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP,tRNAfMetCAU],[MetRS_Met_ATP_tRNAfMetCAU], k_forward= rxn_k['re0000000247_k1']),\n",
    "    Reaction.from_massaction([MetRS_Met_ATP_tRNAfMetCAU],[MetRS_Met_ATP,tRNAfMetCAU], k_forward= rxn_k['re0000000248_k1']),\n",
    "    Reaction.from_massaction([MetRS_MetAMP_PPi,tRNAfMetCAU],[MetRS_MetAMP_PPi_tRNAfMetCAU], k_forward= rxn_k['re0000000249_k1']),\n",
    "    Reaction.from_massaction([MetRS_MetAMP_PPi_tRNAfMetCAU],[MetRS_MetAMP_PPi,tRNAfMetCAU], k_forward= rxn_k['re0000000250_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MetRS_MetAMP,tRNAfMetCAU],[MetRS_MetAMP_tRNAfMetCAU], k_forward= rxn_k['re0000000251_k1']),\n",
    "    Reaction.from_massaction([MetRS_MetAMP_tRNAfMetCAU],[MetRS_MetAMP,tRNAfMetCAU], k_forward= rxn_k['re0000000252_k1']),\n",
    "\n",
    "    Reaction.from_massaction([EFTu_GTP,MettRNAfMetCAU],[EFTu_GTP_MettRNAfMetCAU], k_forward= rxn_k['re0000000288_k1']),\n",
    "    Reaction.from_massaction([EFTu_GTP_MettRNAfMetCAU],[EFTu_GTP,MettRNAfMetCAU], k_forward= rxn_k['re0000000289_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF,MettRNAfMetCAU],[MTF_MettRNAfMetCAU], k_forward= rxn_k['re0000000420_k1']),\n",
    "    Reaction.from_massaction([MTF_MettRNAfMetCAU],[MTF,MettRNAfMetCAU], k_forward= rxn_k['re0000000421_k1']),\n",
    "    Reaction.from_massaction([MTF_FD,MettRNAfMetCAU],[MTF_FD_MettRNAfMetCAU], k_forward= rxn_k['re0000000422_k1']),\n",
    "    Reaction.from_massaction([MTF_FD_MettRNAfMetCAU],[MettRNAfMetCAU,MTF_FD], k_forward= rxn_k['re0000000423_k1']),\n",
    "    Reaction.from_massaction([MTF_MettRNAfMetCAU,FD],[MTF_FD_MettRNAfMetCAU], k_forward= rxn_k['re0000000424_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF_FD_MettRNAfMetCAU],[MTF_MettRNAfMetCAU,FD], k_forward= rxn_k['re0000000425_k1']),\n",
    "    Reaction.from_massaction([MTF_FD_MettRNAfMetCAU],[MTF_THF_fMettRNAfMetCAU], k_forward= rxn_k['re0000000426_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF_THF_fMettRNAfMetCAU],[MTF_THF,fMettRNAfMetCAU], k_forward= rxn_k['re0000000428_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF_THF_fMettRNAfMetCAU],[MTF_fMettRNAfMetCAU,THF], k_forward= rxn_k['re0000000430_k1']),\n",
    "\n",
    "    Reaction.from_massaction([MTF_fMettRNAfMetCAU],[MTF,fMettRNAfMetCAU], k_forward= rxn_k['re0000000432_k1']),\n",
    "\n",
    "    Reaction.from_massaction([IF2_GTP,fMettRNAfMetCAU],[IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000449_k1']),\n",
    "    Reaction.from_massaction([IF2_GTP_fMettRNAfMetCAU],[IF2_GTP,fMettRNAfMetCAU], k_forward= rxn_k['re0000000450_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF3,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000465_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF3,IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000466_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP,fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000467_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP,fMettRNAfMetCAU], k_forward= rxn_k['re0000000468_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA,fMettRNAfMetCAU],[RS30S_IF3_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000473_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_fMettRNAfMetCAU_mRNA],[RS30S_IF3_mRNA,fMettRNAfMetCAU], k_forward= rxn_k['re0000000474_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_mRNA,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000475_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_mRNA,IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000476_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_fMettRNAfMetCAU_mRNA,IF2_GTP],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000477_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_fMettRNAfMetCAU_mRNA,IF2_GTP], k_forward= rxn_k['re0000000478_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_mRNA,fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000479_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_IF2_GTP_mRNA,fMettRNAfMetCAU], k_forward= rxn_k['re0000000480_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU,mRNA],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000483_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU,mRNA], k_forward= rxn_k['re0000000484_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,RS50S],[RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000485_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,RS50S], k_forward= rxn_k['re0000000486_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000497_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF3,IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000498_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP,fMettRNAfMetCAU],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000499_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF3_IF2_GTP,fMettRNAfMetCAU], k_forward= rxn_k['re0000000500_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU,IF1],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000511_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU,IF1], k_forward= rxn_k['re0000000512_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA,fMettRNAfMetCAU],[RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000517_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_mRNA,fMettRNAfMetCAU], k_forward= rxn_k['re0000000518_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_mRNA,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000519_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_mRNA,IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000520_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA,IF2_GTP],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000521_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA,IF2_GTP], k_forward= rxn_k['re0000000522_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_mRNA,fMettRNAfMetCAU],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000523_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_IF2_GTP_mRNA,fMettRNAfMetCAU], k_forward= rxn_k['re0000000524_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU,mRNA],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000527_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU,mRNA], k_forward= rxn_k['re0000000528_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,RS50S],[RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000529_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,RS50S], k_forward= rxn_k['re0000000530_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_fMettRNAfMetCAU_mRNA,IF1],[RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000533_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA],[RS30S_IF3_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000534_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,IF1],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000537_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000538_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,IF1],[RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000539_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000540_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000611_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU],[IF2_GTP_fMettRNAfMetCAU,RS30S], k_forward= rxn_k['re0000000612_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP,fMettRNAfMetCAU],[RS30S_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000613_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU],[fMettRNAfMetCAU,RS30S_IF2_GTP], k_forward= rxn_k['re0000000614_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU,IF3],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000617_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU],[IF3,RS30S_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000618_k1']),\n",
    "    Reaction.from_massaction([RS30S_mRNA,fMettRNAfMetCAU],[RS30S_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000621_k1']),\n",
    "    Reaction.from_massaction([RS30S_fMettRNAfMetCAU_mRNA],[fMettRNAfMetCAU,RS30S_mRNA], k_forward= rxn_k['re0000000622_k1']),\n",
    "    Reaction.from_massaction([RS30S_mRNA,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000623_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF2_GTP_fMettRNAfMetCAU,RS30S_mRNA], k_forward= rxn_k['re0000000624_k1']),\n",
    "    Reaction.from_massaction([RS30S_fMettRNAfMetCAU_mRNA,IF2_GTP],[RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000627_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF2_GTP,RS30S_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000628_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_mRNA,fMettRNAfMetCAU],[RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000629_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA],[fMettRNAfMetCAU,RS30S_IF2_GTP_mRNA], k_forward= rxn_k['re0000000630_k1']),\n",
    "\n",
    "    Reaction.from_massaction([mRNA,RS30S_IF2_GTP_fMettRNAfMetCAU],[RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000633_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF2_GTP_fMettRNAfMetCAU,mRNA], k_forward= rxn_k['re0000000634_k1']),\n",
    "    Reaction.from_massaction([RS30S_fMettRNAfMetCAU_mRNA,IF3],[RS30S_IF3_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000637_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_fMettRNAfMetCAU_mRNA],[IF3,RS30S_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000638_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA,IF3],[RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000641_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF3,RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000642_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000645_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU],[IF2_GTP_fMettRNAfMetCAU,RS30S_IF1], k_forward= rxn_k['re0000000646_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP,fMettRNAfMetCAU],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000647_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU],[fMettRNAfMetCAU,RS30S_IF1_IF2_GTP], k_forward= rxn_k['re0000000648_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF2_GTP_fMettRNAfMetCAU,IF1],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000651_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU],[IF1,RS30S_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000652_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU,IF3],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000655_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU],[IF3,RS30S_IF1_IF2_GTP_fMettRNAfMetCAU], k_forward= rxn_k['re0000000656_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA,fMettRNAfMetCAU],[RS30S_IF1_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000659_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_fMettRNAfMetCAU_mRNA],[fMettRNAfMetCAU,RS30S_IF1_mRNA], k_forward= rxn_k['re0000000660_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_mRNA,fMettRNAfMetCAU],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000661_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA],[fMettRNAfMetCAU,RS30S_IF1_IF2_GTP_mRNA], k_forward= rxn_k['re0000000662_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_fMettRNAfMetCAU_mRNA,IF2_GTP],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000663_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF2_GTP,RS30S_IF1_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000664_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1_mRNA,IF2_GTP_fMettRNAfMetCAU],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000665_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF2_GTP_fMettRNAfMetCAU,RS30S_IF1_mRNA], k_forward= rxn_k['re0000000666_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_fMettRNAfMetCAU_mRNA,IF3],[RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000669_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_fMettRNAfMetCAU_mRNA],[IF3,RS30S_IF1_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000670_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA,IF3],[RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000673_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[IF3,RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000674_k1']),\n",
    "    Reaction.from_massaction([RS30S_fMettRNAfMetCAU_mRNA,IF1],[RS30S_IF1_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000683_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_fMettRNAfMetCAU_mRNA],[RS30S_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000684_k1']),\n",
    "    Reaction.from_massaction([IF1,RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000685_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF2_GTP_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000686_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU,mRNA],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000687_k1']),\n",
    "    Reaction.from_massaction([RS30S_IF1_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS30S_IF1_IF2_GTP_fMettRNAfMetCAU,mRNA], k_forward= rxn_k['re0000000688_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000715_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000716_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA],[RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000717_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA],[RS70S_IF1_IF3_IF2_GTP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000718_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA,IF1],[RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000719_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000720_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA,PO4,], k_forward= rxn_k['re0000000721_k1']), \n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_PO4_fMettRNAfMetCAU_mRNA],[RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA,PO4,], k_forward= rxn_k['re0000000722_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA,IF1],[RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000723_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000724_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF3_fMettRNAfMetCAU_mRNA,IF2_GDP], k_forward= rxn_k['re0000000725_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA,IF3], k_forward= rxn_k['re0000000747_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA,IF2_GDP], k_forward= rxn_k['re0000000749_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA,IF2_GDP],[RS70S_IF1_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000750_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF3_fMettRNAfMetCAU_mRNA,IF2_GDP],[RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000751_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA],[RS70S_IF3_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000753_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF3_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF2_GDP_fMettRNAfMetCAU_mRNA,IF3], k_forward= rxn_k['re0000000755_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF2_GDP_fMettRNAfMetCAU_mRNA,IF1], k_forward= rxn_k['re0000000757_k1']),\n",
    "\n",
    "    Reaction.from_massaction([RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA],[RS70S_IF1_fMettRNAfMetCAU_mRNA,IF2_GDP], k_forward= rxn_k['re0000000759_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_fMettRNAfMetCAU_mRNA,IF2_GDP],[RS70S_IF1_IF2_GDP_fMettRNAfMetCAU_mRNA], k_forward= rxn_k['re0000000760_k1']),\n",
    "    Reaction.from_massaction([RS70S_IF1_IF3_fMettRNAfMetCAU_mRNA],[RS70S_IF1_fMettRNAfMetCAU_mRNA,IF3], k_forward= rxn_k['re0000000761_k1']),\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de509dab-0011-4aa9-8eb4-d16be28a1671",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Addition of AA individual reactions required for desire protein production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1fc45422-aae2-4d27-9009-b1670fcea174",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Creates a list of all the species associated with the aa in the chain, except if aa=Met\n",
    "#Iterates over all the aa in input aa chain\n",
    "for aa in list_AA:\n",
    "    xyz= list_condon[list_AA.index(aa)] #matches the codon for each of the AA\n",
    "    \n",
    "    if aa != 'Met': #does not include Met since equations already exists for it.\n",
    "        asub=aa\n",
    "        asub = Species(asub)\n",
    "        aaAMP = Species(aa+'AMP')\n",
    "        aaRS = Species(aa+'RS')\n",
    "        aaRS_AMP = Species(aa+'RS_AMP')\n",
    "        aaRS_ATP = Species(aa+'RS_ATP')\n",
    "        aaRS_degraded = Species(aa+'RS_degraded')\n",
    "        aaRS_aa= Species(aa+'RS_'+aa)\n",
    "        aaRS_aa_ATP = Species(aa+'RS_'+aa+'_ATP')\n",
    "        aaRS_aaAMP = Species(aa+'RS_'+aa+'AMP')\n",
    "        aaRS_aaAMP_PPi = Species(aa+'RS_'+aa+'AMP_PPi')\n",
    "        \n",
    "        aaRS_AMP_aatRNAaaxyz = Species(aa+'RS_AMP_'+aa+'tRNA'+aa+xyz)\n",
    "        aaRS_ATP_tRNAaaxyz = Species(aa+'RS_ATP_tRNA'+aa+xyz)\n",
    "        aaRS_aa_ATP_tRNAaaxyz = Species(aa+'RS_'+aa+'_ATP_tRNA'+aa+xyz)\n",
    "        aaRS_aa_tRNAaaxyz = Species(aa+'RS_'+aa+'_tRNA'+aa+xyz)\n",
    "        aaRS_aaAMP_PPi_tRNAaaxyz = Species(aa+'RS_'+aa+'AMP_PPi_tRNA'+aa+xyz)\n",
    "        aaRS_aaAMP_tRNAaaxyz = Species(aa+'RS_'+aa+'AMP_tRNA'+aa+xyz)\n",
    "        aaRS_aatRNAaaxyz = Species(aa+'RS_'+aa+'tRNA'+aa+xyz)\n",
    "        aaRS_tRNAaaxyz = Species(aa+'RS_tRNA'+aa+xyz)\n",
    "        aatRNAaaxyz = Species(aa+'tRNA'+aa+xyz)\n",
    "        aatRNAaaxyz_degraded = Species(aa+'tRNA'+aa+xyz+'_degraded')\n",
    "        tRNAaaxyz = Species('tRNA'+aa+xyz)\n",
    "        tRNAaaxyz_degraded = Species('tRNA'+aa+xyz+'_degraded')\n",
    "        \n",
    "        RS50S_tRNAaaxyz= Species('RS50S_tRNA'+aa+xyz)\n",
    "        RS50S_tRNAaaxyz_EFG_GDP = Species('RS50S_tRNA'+aa+xyz+'_EFG_GDP')\n",
    "        RS50S_tRNAaaxyz_RRF = Species('RS50S_tRNA'+aa+xyz+'_RRF')\n",
    "        RS50S_tRNAaaxyz_RRF_EFG_GDP = Species('RS50S_tRNA'+aa+xyz+'_RRF_EFG_GDP')\n",
    "        EFTu_GTP_aatRNAaaxyz = Species('EFTu_GTP_'+aa+'tRNA'+aa+xyz)\n",
    "        \n",
    "        aa = Species(aa)\n",
    "        #######################################################################################################################\n",
    "\n",
    "        #A list of Species\n",
    "        species_aa= [\n",
    "            asub, aaAMP, aaRS, aaRS_AMP, aaRS_ATP,\n",
    "            aaRS_degraded, aaRS_aa, aaRS_aa_ATP, aaRS_aaAMP, \n",
    "            aaRS_aaAMP_PPi, aaRS_AMP_aatRNAaaxyz, aaRS_ATP_tRNAaaxyz, aaRS_aa_ATP_tRNAaaxyz,\n",
    "            aaRS_aa_tRNAaaxyz, aaRS_aaAMP_PPi_tRNAaaxyz, aaRS_aaAMP_tRNAaaxyz, aaRS_aatRNAaaxyz, \n",
    "            aaRS_tRNAaaxyz, aatRNAaaxyz, aatRNAaaxyz_degraded,tRNAaaxyz, tRNAaaxyz_degraded,\n",
    "            \n",
    "            RS50S_tRNAaaxyz, RS50S_tRNAaaxyz_EFG_GDP, RS50S_tRNAaaxyz_RRF, RS50S_tRNAaaxyz_RRF_EFG_GDP, EFTu_GTP_aatRNAaaxyz]        \n",
    "        \n",
    "        #######################################################################################################################\n",
    "        \n",
    "        #Reactions\n",
    "        rxn=[Reaction.from_massaction([aaRS,aa],[aaRS_aa], k_forward = rxn_k['re0000000126_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_PPi],[aaRS_aaAMP,PPi,], k_forward = rxn_k['re0000000127_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_aa],[aaRS,aa], k_forward = rxn_k['re0000000131_k1']),\n",
    "            Reaction.from_massaction([aaRS,ATP],[aaRS_ATP], k_forward = rxn_k['re0000000132_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP],[aaRS,ATP], k_forward = rxn_k['re0000000133_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP,aa],[aaRS_aa_ATP], k_forward = rxn_k['re0000000134_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP],[aaRS_ATP,aa], k_forward = rxn_k['re0000000135_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa,ATP],[aaRS_aa_ATP], k_forward = rxn_k['re0000000136_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP],[aaRS_aa,ATP], k_forward = rxn_k['re0000000137_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_aa_ATP],[aaRS_aaAMP_PPi], k_forward = rxn_k['re0000000140_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_PPi],[aaRS_aa_ATP], k_forward = rxn_k['re0000000141_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_AMP],[aaRS,AMP], k_forward = rxn_k['re0000000145_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_aaAMP],[aaRS,aaAMP], k_forward = rxn_k['re0000000147_k1']),\n",
    "            Reaction.from_massaction([aaRS,aaAMP],[aaRS_aaAMP], k_forward = rxn_k['re0000000148_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_aaAMP_tRNAaaxyz],[aaRS_AMP_aatRNAaaxyz], k_forward = rxn_k['re0000000178_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_AMP_aatRNAaaxyz],[aaRS_aatRNAaaxyz,AMP], k_forward = rxn_k['re0000000180_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_AMP_aatRNAaaxyz],[aatRNAaaxyz,aaRS_AMP], k_forward = rxn_k['re0000000182_k1']),\n",
    "            Reaction.from_massaction([aaRS_AMP,aatRNAaaxyz],[aaRS_AMP_aatRNAaaxyz], k_forward = rxn_k['re0000000183_k1']),\n",
    "            Reaction.from_massaction([aaRS_aatRNAaaxyz],[aaRS,aatRNAaaxyz], k_forward = rxn_k['re0000000184_k1']),\n",
    "            Reaction.from_massaction([aaRS,aatRNAaaxyz],[aaRS_aatRNAaaxyz], k_forward = rxn_k['re0000000185_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS_tRNAaaxyz,aa],[aaRS_aa_tRNAaaxyz], k_forward = rxn_k['re0000000188_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_PPi_tRNAaaxyz],[aaRS_aaAMP_tRNAaaxyz,PPi,], k_forward = rxn_k['re0000000189_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_tRNAaaxyz],[aaRS_tRNAaaxyz,aa], k_forward = rxn_k['re0000000190_k1']),\n",
    "            Reaction.from_massaction([aaRS_tRNAaaxyz,ATP],[aaRS_ATP_tRNAaaxyz], k_forward = rxn_k['re0000000191_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP_tRNAaaxyz],[aaRS_tRNAaaxyz,ATP], k_forward = rxn_k['re0000000192_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP_tRNAaaxyz,aa],[aaRS_aa_ATP_tRNAaaxyz], k_forward = rxn_k['re0000000193_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP_tRNAaaxyz],[aaRS_ATP_tRNAaaxyz,aa], k_forward = rxn_k['re0000000194_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_tRNAaaxyz,ATP],[aaRS_aa_ATP_tRNAaaxyz], k_forward = rxn_k['re0000000195_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP_tRNAaaxyz],[aaRS_aa_tRNAaaxyz,ATP], k_forward = rxn_k['re0000000196_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP_tRNAaaxyz],[aaRS_aaAMP_PPi_tRNAaaxyz], k_forward = rxn_k['re0000000197_k1']),\n",
    "\n",
    "            Reaction.from_massaction([aaRS,tRNAaaxyz],[aaRS_tRNAaaxyz], k_forward = rxn_k['re0000000199_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa,tRNAaaxyz],[aaRS_aa_tRNAaaxyz], k_forward = rxn_k['re0000000200_k1']),\n",
    "            Reaction.from_massaction([aaRS_tRNAaaxyz],[aaRS,tRNAaaxyz], k_forward = rxn_k['re0000000201_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_tRNAaaxyz],[aaRS_aa,tRNAaaxyz], k_forward = rxn_k['re0000000202_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP,tRNAaaxyz],[aaRS_ATP_tRNAaaxyz], k_forward = rxn_k['re0000000203_k1']),\n",
    "            Reaction.from_massaction([aaRS_ATP_tRNAaaxyz],[aaRS_ATP,tRNAaaxyz], k_forward = rxn_k['re0000000204_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP,tRNAaaxyz],[aaRS_aa_ATP_tRNAaaxyz], k_forward = rxn_k['re0000000205_k1']),\n",
    "            Reaction.from_massaction([aaRS_aa_ATP_tRNAaaxyz],[aaRS_aa_ATP,tRNAaaxyz], k_forward = rxn_k['re0000000206_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_PPi,tRNAaaxyz],[aaRS_aaAMP_PPi_tRNAaaxyz], k_forward = rxn_k['re0000000207_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_PPi_tRNAaaxyz],[aaRS_aaAMP_PPi,tRNAaaxyz], k_forward = rxn_k['re0000000208_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP,tRNAaaxyz],[aaRS_aaAMP_tRNAaaxyz], k_forward = rxn_k['re0000000209_k1']),\n",
    "            Reaction.from_massaction([aaRS_aaAMP_tRNAaaxyz],[aaRS_aaAMP,tRNAaaxyz], k_forward = rxn_k['re0000000210_k1']),\n",
    "\n",
    "            Reaction.from_massaction([EFTu_GTP,aatRNAaaxyz],[EFTu_GTP_aatRNAaaxyz], k_forward = rxn_k['re0000000275_k1']),\n",
    "            Reaction.from_massaction([EFTu_GTP_aatRNAaaxyz],[EFTu_GTP,aatRNAaaxyz], k_forward = rxn_k['re0000000276_k1']),\n",
    " \n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_RRF_EFG_GDP],[RS50S_tRNAaaxyz_RRF,EFG_GDP], k_forward = rxn_k['re0000000913_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_RRF_EFG_GDP],[RS50S_tRNAaaxyz_EFG_GDP,RRF], k_forward = rxn_k['re0000000914_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_RRF_EFG_GDP],[RS50S_RRF_EFG_GDP,tRNAaaxyz], k_forward = rxn_k['re0000000915_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_RRF],[RS50S_RRF,tRNAaaxyz], k_forward = rxn_k['re0000000916_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_RRF],[RS50S_tRNAaaxyz,RRF], k_forward = rxn_k['re0000000917_k1']),\n",
    "\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz],[tRNAaaxyz,RS50S], k_forward = rxn_k['re0000000919_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_EFG_GDP],[EFG_GDP,RS50S_tRNAaaxyz], k_forward = rxn_k['re0000000920_k1']),\n",
    "            Reaction.from_massaction([RS50S_tRNAaaxyz_EFG_GDP],[RS50S_EFG_GDP,tRNAaaxyz], k_forward = rxn_k['re0000000921_k1']),\n",
    "            ]\n",
    "\n",
    "        list_of_reactions.append(rxn)\n",
    "        list_species_aa.append(species_aa)\n",
    "    \n",
    "    else:\n",
    "            #Adding MetRS\n",
    "            #This is different than fMet species and reactions for fMet\n",
    "\n",
    "            MetRS_AMP_MettRNAMetCAU = Species('MetRS_AMP_tRNAMetCAU')\n",
    "            MetRS_ATP_tRNAMetCAU = Species('MetRS_ATP_tRNAMetCAU')\n",
    "            MetRS_Met_ATP_tRNAMetCAU = Species('MetRS_Met_ATP_tRNAMetCAU')\n",
    "            MetRS_Met_tRNAMetCAU = Species('MetRS_Met_tRNAMetCAU')\n",
    "            MetRS_MetAMP_PPi_tRNAMetCAU = Species('MetRS_MetAMP_PPi_tRNAMetCAU')\n",
    "            MetRS_MetAMP_tRNAMetCAU = Species('MetRS_MetAMP_tRNAMetCAU')\n",
    "            MetRS_MettRNAMetCAU = Species('MetRS_MettRNAMetCAU')\n",
    "            MetRS_tRNAMetCAU = Species('MetRS_tRNAMetCAU')\n",
    "            MettRNAMetCAU = Species('MettRNAMetCAU')\n",
    "            MettRNAMetCAU_degraded = Species('MettRNAMetCAU_degraded')\n",
    "            tRNAMetCAU = Species('tRNAMetCAU')\n",
    "            tRNAMetCAU_degraded = Species('tRNAMetCAU_degraded')\n",
    "\n",
    "            RS50S_tRNAMetCAU= Species('RS50S_tRNAMetCAU')\n",
    "            RS50S_tRNAMetCAU_EFG_GDP = Species('RS50S_tRNAMetCAU_EFG_GDP')\n",
    "            RS50S_tRNAMetCAU_RRF = Species('RS50S_tRNAMetCAU_RRF')\n",
    "            RS50S_tRNAMetCAU_RRF_EFG_GDP = Species('RS50S_tRNAMetCAU_RRF_EFG_GDP')\n",
    "            EFTu_GTP_MettRNAMetCAU = Species('EFTu_GTP_MettRNAMetCAU')\n",
    "\n",
    "            #######################################################################################################################\n",
    "            species_Met= [MetRS_AMP_MettRNAMetCAU, MetRS_ATP_tRNAMetCAU, MetRS_Met_ATP_tRNAMetCAU,\n",
    "                            MetRS_Met_tRNAMetCAU, MetRS_MetAMP_PPi_tRNAMetCAU, MetRS_MetAMP_tRNAMetCAU, MetRS_MettRNAMetCAU, \n",
    "                            MetRS_tRNAMetCAU, MettRNAMetCAU, MettRNAMetCAU_degraded,tRNAMetCAU, tRNAMetCAU_degraded,\n",
    "\n",
    "                            RS50S_tRNAMetCAU, RS50S_tRNAMetCAU_EFG_GDP, RS50S_tRNAMetCAU_RRF, RS50S_tRNAMetCAU_RRF_EFG_GDP, EFTu_GTP_MettRNAMetCAU]  \n",
    "\n",
    "            #######################################################################################################################\n",
    "            reactions_Met= [\n",
    "                        Reaction.from_massaction([MetRS_MetAMP_tRNAMetCAU],[MetRS_AMP_MettRNAMetCAU], k_forward = rxn_k['re0000000178_k1']),\n",
    "                        Reaction.from_massaction([MetRS_AMP_MettRNAMetCAU],[MetRS_MettRNAMetCAU,AMP], k_forward = rxn_k['re0000000180_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([MetRS_AMP_MettRNAMetCAU],[MettRNAMetCAU,MetRS_AMP], k_forward = rxn_k['re0000000182_k1']),\n",
    "                        Reaction.from_massaction([MetRS_AMP,MettRNAMetCAU],[MetRS_AMP_MettRNAMetCAU], k_forward = rxn_k['re0000000183_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MettRNAMetCAU],[MetRS,MettRNAMetCAU], k_forward = rxn_k['re0000000184_k1']),\n",
    "                        Reaction.from_massaction([MetRS,MettRNAMetCAU],[MetRS_MettRNAMetCAU], k_forward = rxn_k['re0000000185_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([MetRS_tRNAMetCAU,Met],[MetRS_Met_tRNAMetCAU], k_forward = rxn_k['re0000000188_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MetAMP_PPi_tRNAMetCAU],[MetRS_MetAMP_tRNAMetCAU,PPi,], k_forward = rxn_k['re0000000189_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_tRNAMetCAU],[MetRS_tRNAMetCAU,Met], k_forward = rxn_k['re0000000190_k1']),\n",
    "                        Reaction.from_massaction([MetRS_tRNAMetCAU,ATP],[MetRS_ATP_tRNAMetCAU], k_forward = rxn_k['re0000000191_k1']),\n",
    "                        Reaction.from_massaction([MetRS_ATP_tRNAMetCAU],[MetRS_tRNAMetCAU,ATP], k_forward = rxn_k['re0000000192_k1']),\n",
    "                        Reaction.from_massaction([MetRS_ATP_tRNAMetCAU,Met],[MetRS_Met_ATP_tRNAMetCAU], k_forward = rxn_k['re0000000193_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_ATP_tRNAMetCAU],[MetRS_ATP_tRNAMetCAU,Met], k_forward = rxn_k['re0000000194_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_tRNAMetCAU,ATP],[MetRS_Met_ATP_tRNAMetCAU], k_forward = rxn_k['re0000000195_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_ATP_tRNAMetCAU],[MetRS_Met_tRNAMetCAU,ATP], k_forward = rxn_k['re0000000196_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_ATP_tRNAMetCAU],[MetRS_MetAMP_PPi_tRNAMetCAU], k_forward = rxn_k['re0000000197_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([MetRS,tRNAMetCAU],[MetRS_tRNAMetCAU], k_forward = rxn_k['re0000000199_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met,tRNAMetCAU],[MetRS_Met_tRNAMetCAU], k_forward = rxn_k['re0000000200_k1']),\n",
    "                        Reaction.from_massaction([MetRS_tRNAMetCAU],[MetRS,tRNAMetCAU], k_forward = rxn_k['re0000000201_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_tRNAMetCAU],[MetRS_Met,tRNAMetCAU], k_forward = rxn_k['re0000000202_k1']),\n",
    "                        Reaction.from_massaction([MetRS_ATP,tRNAMetCAU],[MetRS_ATP_tRNAMetCAU], k_forward = rxn_k['re0000000203_k1']),\n",
    "                        Reaction.from_massaction([MetRS_ATP_tRNAMetCAU],[MetRS_ATP,tRNAMetCAU], k_forward = rxn_k['re0000000204_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_ATP,tRNAMetCAU],[MetRS_Met_ATP_tRNAMetCAU], k_forward = rxn_k['re0000000205_k1']),\n",
    "                        Reaction.from_massaction([MetRS_Met_ATP_tRNAMetCAU],[MetRS_Met_ATP,tRNAMetCAU], k_forward = rxn_k['re0000000206_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MetAMP_PPi,tRNAMetCAU],[MetRS_MetAMP_PPi_tRNAMetCAU], k_forward = rxn_k['re0000000207_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MetAMP_PPi_tRNAMetCAU],[MetRS_MetAMP_PPi,tRNAMetCAU], k_forward = rxn_k['re0000000208_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MetAMP,tRNAMetCAU],[MetRS_MetAMP_tRNAMetCAU], k_forward = rxn_k['re0000000209_k1']),\n",
    "                        Reaction.from_massaction([MetRS_MetAMP_tRNAMetCAU],[MetRS_MetAMP,tRNAMetCAU], k_forward = rxn_k['re0000000210_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([EFTu_GTP,MettRNAMetCAU],[EFTu_GTP_MettRNAMetCAU], k_forward = rxn_k['re0000000275_k1']),\n",
    "                        Reaction.from_massaction([EFTu_GTP_MettRNAMetCAU],[EFTu_GTP,MettRNAMetCAU], k_forward = rxn_k['re0000000276_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_RRF_EFG_GDP],[RS50S_tRNAMetCAU_RRF,EFG_GDP], k_forward = rxn_k['re0000000913_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_RRF_EFG_GDP],[RS50S_tRNAMetCAU_EFG_GDP,RRF], k_forward = rxn_k['re0000000914_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_RRF_EFG_GDP],[RS50S_RRF_EFG_GDP,tRNAMetCAU], k_forward = rxn_k['re0000000915_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_RRF],[RS50S_RRF,tRNAMetCAU], k_forward = rxn_k['re0000000916_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_RRF],[RS50S_tRNAMetCAU,RRF], k_forward = rxn_k['re0000000917_k1']),\n",
    "\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU],[tRNAMetCAU,RS50S], k_forward = rxn_k['re0000000919_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_EFG_GDP],[EFG_GDP,RS50S_tRNAMetCAU], k_forward = rxn_k['re0000000920_k1']),\n",
    "                        Reaction.from_massaction([RS50S_tRNAMetCAU_EFG_GDP],[RS50S_EFG_GDP,tRNAMetCAU], k_forward = rxn_k['re0000000921_k1']),\n",
    "                        ]\n",
    "            list_of_reactions.append(reactions_Met)\n",
    "            list_species_aa.append(species_Met)\n",
    "list_of_reactions=flatten(list_of_reactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ae29ef7-87e3-4790-a32f-c5ba9611e53e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Compiling TL CRN (CRN_TL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b0fad45e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compile all reactions and species\n",
    "All_species_TL= flatten([list_species_gen,  list_species_fmet,\n",
    "                       flatten(list_species_aa)])\n",
    "\n",
    "All_rxn_TL = flatten([list_of_reaction_gen, list_reaction_fmet,\n",
    "                   list_of_reactions]) \n",
    "        \n",
    "#Buliding the CRN_TL and saving as a SBML file\n",
    "CRN_TL = ChemicalReactionNetwork(species = All_species_TL, reactions = All_rxn_TL)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6d6668c-bea6-44f1-b6e8-eff89d721b9f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Combining all CRNs: CRN_TX + CRN_TL + CRN_linker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2783baaf-3712-4f35-a604-6107582857a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1717, 1044)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Combine_PURE=ChemicalReactionNetwork(species = flatten([CRN_TX.species, CRN_TL.species, ]),#gfp_species\n",
    "                                     reactions = flatten([CRN_TX.reactions, CRN_TL.reactions]))#Rxn_folding,\n",
    "filename='CRN_PURE_MGaptTXwTL'\n",
    "Combine_PURE.write_sbml_file(filename+\".xml\")\n",
    "\n",
    "len(Combine_PURE.reactions),len(Combine_PURE.species)\n",
    "#MGapt TXwTL: (1717, 1044)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "de4bca36-34a1-4fb3-883e-e0e64003ebb0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def update_model_local_to_global_params(filename):\n",
    "    import libsbml\n",
    "    # rename parameters in sbml file so that they have same ID\n",
    "    sbml_doc = libsbml.readSBMLFromFile(filename) # load your old SBML file\n",
    "    param_dict = {}\n",
    "    sbml_model = sbml_doc.getModel()\n",
    "    for rxn in sbml_model.getListOfReactions():\n",
    "        kl = rxn.getKineticLaw()\n",
    "        for param in kl.getListOfLocalParameters():\n",
    "            param_dict[param.getValue()] = param.getId()\n",
    "    added_params = []\n",
    "    changelog = {}\n",
    "    count = 0\n",
    "    for p_val, p in param_dict.items():\n",
    "        sbml_param = sbml_model.createParameter()\n",
    "        if p in added_params:\n",
    "            p_new = p+str(count)\n",
    "            changelog[p_val] = p_new\n",
    "            sbml_param.setId(p_new)\n",
    "            added_params.append(p_new)\n",
    "        else:\n",
    "            sbml_param.setId(p)\n",
    "            added_params.append(p)\n",
    "        count += 1\n",
    "        sbml_param.setValue(p_val)\n",
    "        sbml_param.setConstant(False)\n",
    "\n",
    "    for rxn in sbml_model.getListOfReactions():\n",
    "        kl = rxn.getKineticLaw()\n",
    "        for param in list(kl.getListOfLocalParameters()):\n",
    "            if param.getValue() in changelog:\n",
    "                # print('renaming {0} with {1}'.format(param.getId(), changelog[param.getValue()]))\n",
    "                kl.setFormula(kl.getFormula().replace(param.getId(), changelog[param.getValue()]))\n",
    "            kl.removeLocalParameter(param.getId())\n",
    "\n",
    "\n",
    "    updated_filname = filename.split('.xml')[0] + '_updated.xml'\n",
    "\n",
    "    libsbml.writeSBML(sbml_doc, updated_filname) # write your new SBML\n",
    "    \n",
    "update_model_local_to_global_params(filename+'.xml')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622ef97e-bdc9-4c60-801c-43b78c229e30",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Update model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9915977c-9a93-48b4-9ba0-13a346117c27",
   "metadata": {},
   "source": [
    "# Computing environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f3c669e6-0904-4e3f-a847-3f36de416973",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python implementation: CPython\n",
      "Python version       : 3.8.17\n",
      "IPython version      : 8.12.2\n",
      "\n",
      "bioscrape  : 1.2.1\n",
      "bokeh      : 2.4.0\n",
      "panel      : 0.13.1\n",
      "jupyterlab : 3.6.5\n",
      "biocrnpyler: 1.1.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -p bioscrape,bokeh,panel,jupyterlab,biocrnpyler"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
